// @niivue/niivue/build/drawing/index.js
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);

// src/logger.ts
var _Log = class _Log {
  constructor({ name = "niivue", level = "info" } = {}) {
    __publicField(this, "level");
    __publicField(this, "name");
    this.name = `${name}`;
    this.level = level;
  }
  debug(...args) {
    if (_Log.levels[this.level] > _Log.levels.debug) {
      return;
    }
    console.debug(`${this.name}-debug`, ...args);
  }
  info(...args) {
    if (_Log.levels[this.level] > _Log.levels.info) {
      return;
    }
    console.info(`${this.name}-info`, ...args);
  }
  warn(...args) {
    if (_Log.levels[this.level] > _Log.levels.warn) {
      return;
    }
    console.warn(`${this.name}-warn`, ...args);
  }
  error(...args) {
    if (_Log.levels[this.level] > _Log.levels.error) {
      return;
    }
    console.error(`${this.name}-error`, ...args);
  }
  fatal(...args) {
    if (_Log.levels[this.level] > _Log.levels.fatal) {
      return;
    }
    console.error(`${this.name}-fatal`, ...args);
  }
  setLogLevel(level) {
    this.level = level;
  }
  setName(name) {
    this.name = name;
  }
};
// map 'debug' 'info' 'warn' 'error' 'fatal' 'silent' to numbers
// for comparison
__publicField(_Log, "levels", {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3,
  fatal: 4,
  silent: Infinity
});
var Log = _Log;
var log = new Log({ name: "niivue", level: "info" });

// src/drawing/rle.ts
function encodeRLE(data) {
  const dl = data.length;
  let dp = 0;
  const r = new Uint8Array(dl + Math.ceil(0.01 * dl));
  const rI = new Int8Array(r.buffer);
  let rp = 0;
  while (dp < dl) {
    let v = data[dp];
    dp++;
    let rl = 1;
    while (rl < 129 && dp < dl && data[dp] === v) {
      dp++;
      rl++;
    }
    if (rl > 1) {
      rI[rp] = -rl + 1;
      rp++;
      r[rp] = v;
      rp++;
      continue;
    }
    while (dp < dl) {
      if (rl > 127) {
        break;
      }
      if (dp + 2 < dl) {
        if (v !== data[dp] && data[dp + 2] === data[dp] && data[dp + 1] === data[dp]) {
          break;
        }
      }
      v = data[dp];
      dp++;
      rl++;
    }
    r[rp] = rl - 1;
    rp++;
    for (let i = 0; i < rl; i++) {
      r[rp] = data[dp - rl + i];
      rp++;
    }
  }
  log.debug("PackBits " + dl + " -> " + rp + " bytes (x" + dl / rp + ")");
  return r.slice(0, rp);
}
function decodeRLE(rle, decodedlen) {
  const r = new Uint8Array(rle.buffer);
  const rI = new Int8Array(r.buffer);
  let rp = 0;
  const d = new Uint8Array(decodedlen);
  let dp = 0;
  while (rp < r.length) {
    const hdr = rI[rp];
    rp++;
    if (hdr < 0) {
      const v = rI[rp];
      rp++;
      for (let i = 0; i < 1 - hdr; i++) {
        d[dp] = v;
        dp++;
      }
    } else {
      for (let i = 0; i < hdr + 1; i++) {
        d[dp] = rI[rp];
        rp++;
        dp++;
      }
    }
  }
  return d;
}

// src/drawing/undo.ts
var drawUndo = ({ drawUndoBitmaps, currentDrawUndoBitmap, drawBitmap }) => {
  const len = drawUndoBitmaps.length;
  if (len < 1) {
    log.debug("undo bitmaps not loaded");
    return;
  }
  currentDrawUndoBitmap--;
  if (currentDrawUndoBitmap < 0) {
    currentDrawUndoBitmap = len - 1;
  }
  if (currentDrawUndoBitmap >= len) {
    currentDrawUndoBitmap = 0;
  }
  if (drawUndoBitmaps[currentDrawUndoBitmap].length < 2) {
    log.debug("drawUndo is misbehaving");
    return;
  }
  drawBitmap = decodeRLE(drawUndoBitmaps[currentDrawUndoBitmap], drawBitmap.length);
  return { drawBitmap, currentDrawUndoBitmap };
};

// src/nvdocument.ts
import { vec3 as vec313, vec4 as vec49 } from "gl-matrix";

// src/nvserializer.ts
import { vec3 as vec312 } from "gl-matrix";

// src/nvutilities.ts
import arrayEqual from "array-equal";
import { mat4, vec3, vec4 } from "gl-matrix";

// src/nvimage/index.ts
import { readHeaderAsync as readHeaderAsync4 } from "nifti-reader-js";
import { v4 as uuidv42 } from "@lukeed/uuid";

// src/nvimage/utils.ts
import { vec3 as vec32 } from "gl-matrix";
var ImageType = /* @__PURE__ */ ((ImageType3) => {
  ImageType3[ImageType3["UNKNOWN"] = 0] = "UNKNOWN";
  ImageType3[ImageType3["NII"] = 1] = "NII";
  ImageType3[ImageType3["DCM"] = 2] = "DCM";
  ImageType3[ImageType3["DCM_MANIFEST"] = 3] = "DCM_MANIFEST";
  ImageType3[ImageType3["MIH"] = 4] = "MIH";
  ImageType3[ImageType3["MIF"] = 5] = "MIF";
  ImageType3[ImageType3["NHDR"] = 6] = "NHDR";
  ImageType3[ImageType3["NRRD"] = 7] = "NRRD";
  ImageType3[ImageType3["MHD"] = 8] = "MHD";
  ImageType3[ImageType3["MHA"] = 9] = "MHA";
  ImageType3[ImageType3["MGH"] = 10] = "MGH";
  ImageType3[ImageType3["MGZ"] = 11] = "MGZ";
  ImageType3[ImageType3["V"] = 12] = "V";
  ImageType3[ImageType3["V16"] = 13] = "V16";
  ImageType3[ImageType3["VMR"] = 14] = "VMR";
  ImageType3[ImageType3["HEAD"] = 15] = "HEAD";
  ImageType3[ImageType3["DCM_FOLDER"] = 16] = "DCM_FOLDER";
  ImageType3[ImageType3["SRC"] = 17] = "SRC";
  ImageType3[ImageType3["FIB"] = 18] = "FIB";
  ImageType3[ImageType3["BMP"] = 19] = "BMP";
  ImageType3[ImageType3["ZARR"] = 20] = "ZARR";
  ImageType3[ImageType3["NPY"] = 21] = "NPY";
  ImageType3[ImageType3["NPZ"] = 22] = "NPZ";
  ImageType3[ImageType3["HDR"] = 23] = "HDR";
  return ImageType3;
})(ImageType || {});
var NVIMAGE_TYPE = Object.freeze({
  ...ImageType,
  parse: (ext) => {
    let imageType = 0 /* UNKNOWN */;
    switch (ext.toUpperCase()) {
      case "":
      case "DCM":
        imageType = 2 /* DCM */;
        break;
      case "TXT":
        imageType = 3 /* DCM_MANIFEST */;
        break;
      case "FZ":
      case "GQI":
      case "QSDR":
      case "FIB":
        imageType = 18 /* FIB */;
        break;
      case "HDR":
      case "NII":
        imageType = 1 /* NII */;
        break;
      case "MIH":
        imageType = 4 /* MIH */;
        break;
      case "MIF":
        imageType = 5 /* MIF */;
        break;
      case "NHDR":
        imageType = 6 /* NHDR */;
        break;
      case "NRRD":
        imageType = 7 /* NRRD */;
        break;
      case "MHD":
        imageType = 8 /* MHD */;
        break;
      case "MHA":
        imageType = 9 /* MHA */;
        break;
      case "MGH":
        imageType = 10 /* MGH */;
        break;
      case "MGZ":
        imageType = 11 /* MGZ */;
        break;
      case "NPY":
        imageType = 21 /* NPY */;
        break;
      case "NPZ":
        imageType = 22 /* NPZ */;
        break;
      case "SRC":
        imageType = 17 /* SRC */;
        break;
      case "V":
        imageType = 12 /* V */;
        break;
      case "V16":
        imageType = 13 /* V16 */;
        break;
      case "VMR":
        imageType = 14 /* VMR */;
        break;
      case "HEAD":
        imageType = 15 /* HEAD */;
        break;
      case "PNG":
      case "BMP":
      case "GIF":
      case "JPG":
      case "JPEG":
        imageType = 19 /* BMP */;
        break;
      case "ZARR":
        imageType = 20 /* ZARR */;
        break;
    }
    return imageType;
  }
});

// src/nvimage/ImageWriter.ts
import { NIFTI1 } from "nifti-reader-js";

// src/nvimage/VolumeUtils.ts
import { vec4 as vec42 } from "gl-matrix";

// src/nvimage/ImageReaders/afni.ts
import { NIFTI1 as NIFTI12, NIFTIEXTENSION } from "nifti-reader-js";
import { mat4 as mat42, vec3 as vec33, vec4 as vec43 } from "gl-matrix";

// src/nvimage/ImageReaders/brainvoyager.ts
import { NIFTI1 as NIFTI13 } from "nifti-reader-js";

// src/nvimage/ImageReaders/dsistudio.ts
import { NIFTI1 as NIFTI14 } from "nifti-reader-js";

// src/nvimage/ImageReaders/ecat.ts
import { NIFTI1 as NIFTI15 } from "nifti-reader-js";

// src/nvimage/ImageReaders/image.ts
import { NIFTI1 as NIFTI16 } from "nifti-reader-js";

// src/nvimage/ImageReaders/itk.ts
import { NIFTI1 as NIFTI17 } from "nifti-reader-js";
import { mat3, vec3 as vec34 } from "gl-matrix";

// src/nvimage/ImageReaders/mgh.ts
import { NIFTI1 as NIFTI18 } from "nifti-reader-js";
import { mat4 as mat43 } from "gl-matrix";

// src/nvimage/ImageReaders/mrtrix.ts
import { NIFTI1 as NIFTI19 } from "nifti-reader-js";

// src/nvimage/ImageReaders/nii.ts
import { isCompressed, decompressAsync, readHeaderAsync, readImage, hasExtension } from "nifti-reader-js";

// src/nvimage/ImageReaders/nrrd.ts
import { NIFTI1 as NIFTI110 } from "nifti-reader-js";
import { mat3 as mat32, mat4 as mat44, vec3 as vec35 } from "gl-matrix";

// src/nvimage/ImageReaders/numpy.ts
import { NIFTI1 as NIFTI111 } from "nifti-reader-js";

// src/nvimage/ImageReaders/zarr.ts
import { NIFTI1 as NIFTI112 } from "nifti-reader-js";

// src/nvimage/CoordinateTransform.ts
import { mat4 as mat45, vec3 as vec36, vec4 as vec44 } from "gl-matrix";

// src/nvimage/ImageOrientation.ts
import { mat3 as mat33, mat4 as mat46, vec3 as vec37, vec4 as vec45 } from "gl-matrix";
import { readHeaderAsync as readHeaderAsync2 } from "nifti-reader-js";

// src/cmaps/index.ts
var cmaps_exports = {};
__export(cmaps_exports, {
  $itksnap: () => itksnap_default,
  $slicer3d: () => slicer3d_default,
  actc: () => actc_default,
  afni_blues_inv: () => afni_blues_inv_default,
  afni_reds_inv: () => afni_reds_inv_default,
  batlow: () => batlow_default,
  bcgwhw: () => bcgwhw_default,
  bcgwhw_dark: () => bcgwhw_dark_default,
  blue: () => blue_default,
  blue2cyan: () => blue2cyan_default,
  blue2magenta: () => blue2magenta_default,
  blue2red: () => blue2red_default,
  bluegrn: () => bluegrn_default,
  bone: () => bone_default,
  bronze: () => bronze_default,
  cet_l17: () => cet_l17_default,
  cividis: () => cividis_default,
  cool: () => cool_default,
  copper: () => copper_default,
  copper2: () => copper2_default,
  ct_airways: () => ct_airways_default,
  ct_artery: () => ct_artery_default,
  ct_bones: () => ct_bones_default,
  ct_brain: () => ct_brain_default,
  ct_brain_gray: () => ct_brain_gray_default,
  ct_cardiac: () => ct_cardiac_default,
  ct_head: () => ct_head_default,
  ct_kidneys: () => ct_kidneys_default,
  ct_liver: () => ct_liver_default,
  ct_muscles: () => ct_muscles_default,
  ct_scalp: () => ct_scalp_default,
  ct_skull: () => ct_skull_default,
  ct_soft: () => ct_soft_default,
  ct_soft_tissue: () => ct_soft_tissue_default,
  ct_surface: () => ct_surface_default,
  ct_vessels: () => ct_vessels_default,
  ct_w_contrast: () => ct_w_contrast_default,
  cubehelix: () => cubehelix_default,
  electric_blue: () => electric_blue_default,
  freesurfer: () => freesurfer_default,
  ge_color: () => ge_color_default,
  gold: () => gold_default,
  gray: () => gray_default,
  green: () => green_default,
  green2cyan: () => green2cyan_default,
  green2orange: () => green2orange_default,
  hot: () => hot_default,
  hotiron: () => hotiron_default,
  hsv: () => hsv_default,
  inferno: () => inferno_default,
  jet: () => jet_default,
  kry: () => kry_default,
  linspecer: () => linspecer_default,
  lipari: () => lipari_default,
  magma: () => magma_default,
  mako: () => mako_default,
  navia: () => navia_default,
  nih: () => nih_default,
  plasma: () => plasma_default,
  random: () => random_default,
  red: () => red_default,
  redyell: () => redyell_default,
  rocket: () => rocket_default,
  roi_i256: () => roi_i256_default,
  surface: () => surface_default,
  thermal: () => thermal_default,
  turbo: () => turbo_default,
  violet: () => violet_default,
  viridis: () => viridis_default,
  warm: () => warm_default,
  winter: () => winter_default,
  x_rain: () => x_rain_default
});

// src/cmaps/_itksnap.json
var itksnap_default = {
  R: [
    0,
    255,
    0,
    0,
    255,
    0,
    255,
    255,
    0,
    205,
    210,
    102,
    0,
    0,
    46,
    255,
    106,
    221,
    233,
    165,
    255,
    147,
    218,
    75,
    255,
    60,
    255,
    255,
    218,
    0,
    188,
    255,
    255,
    222,
    127,
    139,
    124,
    255,
    70,
    0,
    238,
    238,
    240,
    245,
    184,
    32,
    255,
    25,
    112,
    34,
    248,
    245,
    255,
    144,
    173,
    65,
    255,
    250,
    128,
    50,
    244,
    255,
    123,
    255,
    173,
    255,
    127,
    255,
    143,
    220,
    253,
    255,
    0,
    0,
    128,
    255,
    250,
    148,
    178,
    255,
    135,
    100,
    240,
    250,
    255,
    107,
    135,
    0,
    139,
    245,
    186,
    255,
    255,
    0,
    210,
    255,
    47,
    72,
    175,
    128,
    176,
    255,
    139,
    240,
    255,
    216,
    119,
    219,
    72,
    255,
    199,
    154,
    189,
    240,
    230,
    0,
    85,
    64,
    153,
    205,
    250,
    95,
    0,
    255,
    224,
    176,
    138,
    30,
    240,
    152,
    160
  ],
  G: [
    0,
    0,
    255,
    0,
    255,
    255,
    0,
    239,
    0,
    133,
    180,
    205,
    0,
    139,
    139,
    228,
    90,
    160,
    150,
    42,
    250,
    112,
    112,
    0,
    182,
    179,
    235,
    228,
    165,
    128,
    143,
    105,
    218,
    184,
    255,
    69,
    252,
    255,
    130,
    100,
    130,
    232,
    255,
    222,
    134,
    178,
    20,
    25,
    128,
    139,
    248,
    255,
    160,
    238,
    255,
    105,
    99,
    240,
    0,
    205,
    164,
    255,
    104,
    165,
    216,
    192,
    255,
    140,
    188,
    20,
    245,
    250,
    206,
    255,
    0,
    250,
    128,
    0,
    34,
    127,
    206,
    149,
    230,
    235,
    245,
    142,
    206,
    0,
    0,
    245,
    85,
    228,
    222,
    191,
    105,
    248,
    79,
    61,
    238,
    128,
    224,
    240,
    0,
    255,
    215,
    191,
    136,
    112,
    209,
    0,
    21,
    205,
    183,
    248,
    230,
    250,
    107,
    224,
    50,
    92,
    250,
    158,
    128,
    69,
    255,
    196,
    43,
    144,
    128,
    251,
    82
  ],
  B: [
    0,
    0,
    0,
    255,
    0,
    255,
    255,
    213,
    205,
    63,
    140,
    170,
    128,
    139,
    87,
    225,
    205,
    221,
    122,
    42,
    250,
    219,
    214,
    130,
    193,
    113,
    205,
    196,
    32,
    128,
    143,
    180,
    185,
    135,
    0,
    19,
    0,
    224,
    180,
    0,
    238,
    170,
    240,
    179,
    11,
    170,
    147,
    112,
    144,
    34,
    255,
    250,
    122,
    144,
    47,
    225,
    71,
    230,
    0,
    50,
    96,
    240,
    238,
    0,
    230,
    203,
    212,
    0,
    143,
    60,
    230,
    240,
    209,
    127,
    128,
    205,
    114,
    211,
    34,
    80,
    235,
    237,
    140,
    215,
    238,
    35,
    250,
    139,
    139,
    220,
    211,
    181,
    173,
    255,
    30,
    220,
    79,
    139,
    238,
    0,
    230,
    245,
    0,
    255,
    0,
    216,
    153,
    147,
    204,
    255,
    133,
    50,
    107,
    255,
    250,
    154,
    47,
    208,
    204,
    92,
    210,
    160,
    0,
    0,
    255,
    222,
    226,
    255,
    128,
    152,
    45
  ],
  A: [
    0,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255
  ],
  I: [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30,
    31,
    32,
    33,
    34,
    35,
    36,
    37,
    38,
    39,
    40,
    41,
    42,
    43,
    44,
    45,
    46,
    47,
    48,
    49,
    50,
    51,
    52,
    53,
    54,
    55,
    56,
    57,
    58,
    59,
    60,
    61,
    62,
    63,
    64,
    65,
    66,
    67,
    68,
    69,
    70,
    71,
    72,
    73,
    74,
    75,
    76,
    77,
    78,
    79,
    80,
    81,
    82,
    83,
    84,
    85,
    86,
    87,
    88,
    89,
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
    99,
    100,
    101,
    102,
    103,
    104,
    105,
    106,
    107,
    108,
    109,
    110,
    111,
    112,
    113,
    114,
    115,
    116,
    117,
    118,
    119,
    120,
    121,
    122,
    123,
    124,
    125,
    126,
    127,
    128,
    129,
    130
  ]
};

// src/cmaps/_slicer3d.json
var slicer3d_default = {
  labels: [
    "background",
    "tissue",
    "bone",
    "skin",
    "connective tissue",
    "blood",
    "organ",
    "mass",
    "muscle",
    "foreign object",
    "waste",
    "teeth",
    "fat",
    "gray matter",
    "white matter",
    "nerve",
    "vein",
    "artery",
    "capillary",
    "ligament",
    "tendon",
    "cartilage",
    "meniscus",
    "lymph node",
    "lymphatic vessel",
    "cerebro-spinal fluid",
    "bile",
    "urine",
    "feces",
    "gas",
    "fluid",
    "edema",
    "bleeding",
    "necrosis",
    "clot",
    "embolism",
    "head",
    "central nervous system",
    "brain",
    "gray matter of brain",
    "telencephalon",
    "cerebral cortex",
    "right frontal lobe",
    "left frontal lobe",
    "right temporal lobe",
    "left temporal lobe",
    "right parietal lobe",
    "left parietal lobe",
    "right occipital lobe",
    "left occipital lobe",
    "right insular lobe",
    "left insular lobe",
    "right limbic lobe",
    "left limbic lobe",
    "right striatum",
    "left striatum",
    "right caudate nucleus",
    "left caudate nucleus",
    "right putamen",
    "left putamen",
    "right pallidum",
    "left pallidum",
    "right amygdaloid complex",
    "left amygdaloid complex",
    "diencephalon",
    "thalamus",
    "right thalamus",
    "left thalamus",
    "pineal gland",
    "midbrain",
    "substantia nigra",
    "right substantia nigra",
    "left substantia nigra",
    "cerebral white matter",
    "right superior longitudinal fasciculus",
    "left superior longitudinal fasciculus",
    "right inferior longitudinal fasciculus",
    "left inferior longitudinal fasciculus",
    "right arcuate fasciculus",
    "left arcuate fasciculus",
    "right uncinate fasciculus",
    "left uncinate fasciculus",
    "right cingulum bundle",
    "left cingulum bundle",
    "projection fibers",
    "right corticospinal tract",
    "left corticospinal tract",
    "right optic radiation",
    "left optic radiation",
    "right medial lemniscus",
    "left medial lemniscus",
    "right superior cerebellar peduncle",
    "left superior cerebellar peduncle",
    "right middle cerebellar peduncle",
    "left middle cerebellar peduncle",
    "right inferior cerebellar peduncle",
    "left inferior cerebellar peduncle",
    "optic chiasm",
    "right optic tract",
    "left optic tract",
    "right fornix",
    "left fornix",
    "commissural fibers",
    "corpus callosum",
    "posterior commissure",
    "cerebellar white matter",
    "CSF space",
    "ventricles of brain",
    "right lateral ventricle",
    "left lateral ventricle",
    "right third ventricle",
    "left third ventricle",
    "cerebral aqueduct",
    "fourth ventricle",
    "subarachnoid space",
    "spinal cord",
    "gray matter of spinal cord",
    "white matter of spinal cord",
    "endocrine system of brain",
    "pituitary gland",
    "adenohypophysis",
    "neurohypophysis",
    "meninges",
    "dura mater",
    "arachnoid",
    "pia mater",
    "muscles of head",
    "salivary glands",
    "lips",
    "nose",
    "tongue",
    "soft palate",
    "right inner ear",
    "left inner ear",
    "right external ear",
    "left external ear",
    "right middle ear",
    "left middle ear",
    "right eyeball",
    "left eyeball",
    "skull",
    "right frontal bone",
    "left frontal bone",
    "right parietal bone",
    "left parietal bone",
    "right temporal bone",
    "left temporal bone",
    "right sphenoid bone",
    "left sphenoid bone",
    "right ethmoid bone",
    "left ethmoid bone",
    "occipital bone",
    "maxilla",
    "right zygomatic bone",
    "right lacrimal bone",
    "vomer bone",
    "right palatine bone",
    "left palatine bone",
    "mandible",
    "neck",
    "muscles of neck",
    "pharynx",
    "larynx",
    "thyroid gland",
    "right parathyroid glands",
    "left parathyroid glands",
    "skeleton of neck",
    "hyoid bone",
    "cervical vertebral column",
    "thorax",
    "trachea",
    "bronchi",
    "right lung",
    "left lung",
    "superior lobe of right lung",
    "superior lobe of left lung",
    "middle lobe of right lung",
    "inferior lobe of right lung",
    "inferior lobe of left lung",
    "pleura",
    "heart",
    "right atrium",
    "left atrium",
    "atrial septum",
    "ventricular septum",
    "right ventricle of heart",
    "left ventricle of heart",
    "mitral valve",
    "tricuspid valve",
    "aortic valve",
    "pulmonary valve",
    "aorta",
    "pericardium",
    "pericardial cavity",
    "esophagus",
    "thymus",
    "mediastinum",
    "skin of thoracic wall",
    "muscles of thoracic wall",
    "skeleton of thorax",
    "thoracic vertebral column",
    "ribs",
    "sternum",
    "right clavicle",
    "left clavicle",
    "abdominal cavity",
    "abdomen",
    "peritoneum",
    "omentum",
    "peritoneal cavity",
    "retroperitoneal space",
    "stomach",
    "duodenum",
    "small bowel",
    "colon",
    "anus",
    "liver",
    "biliary tree",
    "gallbladder",
    "pancreas",
    "spleen",
    "urinary system",
    "right kidney",
    "left kidney",
    "right ureter",
    "left ureter",
    "urinary bladder",
    "urethra",
    "right adrenal gland",
    "left adrenal gland",
    "female internal genitalia",
    "uterus",
    "right fallopian tube",
    "left fallopian tube",
    "right ovary",
    "left ovary",
    "vagina",
    "male internal genitalia",
    "prostate",
    "right seminal vesicle",
    "left seminal vesicle",
    "right deferent duct",
    "left deferent duct",
    "skin of abdominal wall",
    "muscles of abdominal wall",
    "skeleton of abdomen",
    "lumbar vertebral column",
    "female external genitalia",
    "male external genitalia",
    "skeleton of upper limb",
    "muscles of upper limb",
    "right upper limb",
    "left upper limb",
    "right shoulder",
    "left shoulder",
    "right arm"
  ],
  R: [
    0,
    128,
    241,
    177,
    111,
    216,
    221,
    144,
    192,
    220,
    78,
    255,
    230,
    200,
    250,
    244,
    0,
    216,
    183,
    183,
    152,
    111,
    178,
    68,
    111,
    85,
    0,
    214,
    78,
    218,
    170,
    140,
    188,
    216,
    145,
    150,
    177,
    244,
    250,
    200,
    68,
    128,
    83,
    83,
    162,
    162,
    141,
    141,
    182,
    182,
    188,
    188,
    154,
    154,
    177,
    177,
    30,
    30,
    210,
    210,
    48,
    48,
    98,
    98,
    69,
    166,
    122,
    122,
    253,
    145,
    46,
    0,
    0,
    250,
    127,
    127,
    159,
    159,
    125,
    125,
    106,
    106,
    154,
    154,
    126,
    201,
    201,
    78,
    78,
    174,
    174,
    139,
    139,
    148,
    148,
    186,
    186,
    99,
    156,
    156,
    64,
    64,
    138,
    97,
    126,
    194,
    85,
    88,
    88,
    88,
    88,
    88,
    88,
    88,
    88,
    244,
    200,
    250,
    82,
    57,
    60,
    92,
    255,
    255,
    255,
    255,
    201,
    70,
    188,
    177,
    166,
    182,
    229,
    229,
    174,
    174,
    201,
    201,
    194,
    194,
    241,
    203,
    203,
    229,
    229,
    255,
    255,
    209,
    209,
    248,
    248,
    255,
    196,
    255,
    255,
    255,
    242,
    242,
    222,
    177,
    213,
    184,
    150,
    62,
    62,
    62,
    242,
    250,
    255,
    177,
    182,
    175,
    197,
    197,
    172,
    172,
    202,
    224,
    224,
    255,
    206,
    210,
    203,
    233,
    195,
    181,
    152,
    159,
    166,
    218,
    225,
    224,
    255,
    184,
    211,
    47,
    255,
    173,
    188,
    255,
    226,
    253,
    244,
    205,
    205,
    186,
    177,
    255,
    234,
    204,
    180,
    216,
    255,
    205,
    204,
    255,
    221,
    0,
    139,
    249,
    157,
    203,
    185,
    185,
    247,
    247,
    222,
    124,
    249,
    249,
    244,
    255,
    255,
    227,
    213,
    213,
    193,
    216,
    230,
    245,
    245,
    241,
    241,
    177,
    171,
    217,
    212,
    185,
    185,
    198,
    194,
    177,
    177,
    177,
    177,
    177
  ],
  G: [
    0,
    174,
    214,
    122,
    184,
    101,
    130,
    238,
    104,
    245,
    63,
    250,
    220,
    200,
    250,
    214,
    151,
    101,
    156,
    214,
    189,
    184,
    212,
    172,
    197,
    188,
    145,
    230,
    63,
    255,
    250,
    224,
    65,
    191,
    60,
    98,
    122,
    214,
    250,
    200,
    131,
    174,
    146,
    146,
    115,
    115,
    93,
    93,
    166,
    166,
    135,
    135,
    150,
    150,
    140,
    140,
    111,
    111,
    157,
    157,
    129,
    129,
    153,
    153,
    110,
    113,
    101,
    101,
    135,
    92,
    101,
    108,
    108,
    250,
    150,
    150,
    116,
    116,
    102,
    102,
    174,
    174,
    146,
    146,
    126,
    160,
    160,
    152,
    152,
    140,
    140,
    126,
    126,
    120,
    120,
    135,
    135,
    106,
    171,
    171,
    123,
    123,
    95,
    113,
    161,
    195,
    188,
    106,
    106,
    106,
    106,
    106,
    106,
    106,
    106,
    214,
    200,
    250,
    174,
    157,
    143,
    162,
    244,
    244,
    244,
    244,
    121,
    163,
    91,
    122,
    84,
    105,
    147,
    147,
    122,
    122,
    112,
    112,
    142,
    142,
    213,
    179,
    179,
    204,
    204,
    243,
    243,
    185,
    185,
    223,
    223,
    230,
    172,
    255,
    250,
    237,
    217,
    217,
    198,
    122,
    124,
    105,
    208,
    162,
    162,
    162,
    206,
    210,
    255,
    122,
    228,
    216,
    165,
    165,
    138,
    138,
    164,
    186,
    186,
    245,
    110,
    115,
    108,
    138,
    100,
    85,
    55,
    63,
    70,
    123,
    130,
    97,
    244,
    122,
    171,
    150,
    244,
    121,
    95,
    239,
    202,
    232,
    217,
    179,
    179,
    124,
    122,
    255,
    234,
    142,
    119,
    132,
    253,
    167,
    168,
    224,
    130,
    145,
    150,
    180,
    108,
    136,
    102,
    102,
    182,
    182,
    154,
    186,
    186,
    186,
    170,
    181,
    190,
    153,
    141,
    141,
    123,
    146,
    158,
    172,
    172,
    172,
    172,
    124,
    85,
    198,
    188,
    135,
    135,
    175,
    98,
    122,
    122,
    122,
    122,
    122
  ],
  B: [
    0,
    128,
    145,
    101,
    210,
    79,
    101,
    144,
    88,
    20,
    0,
    220,
    70,
    235,
    210,
    49,
    206,
    79,
    220,
    211,
    207,
    210,
    242,
    100,
    131,
    255,
    30,
    130,
    0,
    255,
    250,
    228,
    28,
    216,
    66,
    83,
    101,
    49,
    225,
    215,
    98,
    128,
    164,
    164,
    105,
    105,
    137,
    137,
    110,
    110,
    166,
    166,
    201,
    201,
    190,
    190,
    85,
    85,
    166,
    166,
    126,
    126,
    112,
    112,
    53,
    137,
    38,
    38,
    192,
    109,
    131,
    112,
    112,
    225,
    88,
    88,
    163,
    163,
    154,
    154,
    155,
    155,
    83,
    83,
    55,
    133,
    133,
    141,
    141,
    103,
    103,
    177,
    177,
    72,
    72,
    135,
    135,
    24,
    108,
    108,
    147,
    147,
    74,
    158,
    197,
    164,
    255,
    215,
    215,
    215,
    215,
    215,
    215,
    215,
    215,
    49,
    215,
    225,
    128,
    110,
    83,
    109,
    209,
    209,
    209,
    209,
    77,
    117,
    95,
    101,
    94,
    107,
    118,
    118,
    90,
    90,
    73,
    73,
    0,
    0,
    144,
    77,
    77,
    109,
    109,
    152,
    152,
    85,
    85,
    131,
    131,
    138,
    68,
    167,
    160,
    145,
    123,
    123,
    101,
    101,
    109,
    108,
    243,
    114,
    114,
    114,
    142,
    139,
    207,
    101,
    255,
    244,
    145,
    145,
    115,
    115,
    140,
    162,
    162,
    217,
    84,
    89,
    81,
    112,
    73,
    57,
    13,
    27,
    38,
    97,
    104,
    76,
    209,
    154,
    143,
    103,
    209,
    88,
    76,
    172,
    134,
    158,
    154,
    108,
    108,
    161,
    101,
    220,
    194,
    178,
    153,
    105,
    229,
    142,
    143,
    199,
    101,
    30,
    98,
    111,
    162,
    116,
    83,
    83,
    164,
    164,
    132,
    223,
    150,
    150,
    147,
    158,
    165,
    130,
    113,
    113,
    103,
    127,
    140,
    147,
    147,
    151,
    151,
    92,
    68,
    131,
    102,
    134,
    134,
    125,
    79,
    101,
    101,
    101,
    101,
    101
  ],
  A: [
    0,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255
  ],
  I: [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30,
    31,
    32,
    33,
    34,
    35,
    36,
    37,
    38,
    39,
    40,
    41,
    42,
    43,
    44,
    45,
    46,
    47,
    48,
    49,
    50,
    51,
    52,
    53,
    54,
    55,
    56,
    57,
    58,
    59,
    60,
    61,
    62,
    63,
    64,
    65,
    66,
    67,
    68,
    69,
    70,
    71,
    72,
    73,
    74,
    75,
    76,
    77,
    78,
    79,
    80,
    81,
    82,
    83,
    84,
    85,
    86,
    87,
    88,
    89,
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
    99,
    100,
    101,
    102,
    103,
    104,
    105,
    106,
    107,
    108,
    109,
    110,
    111,
    112,
    113,
    114,
    115,
    116,
    117,
    118,
    119,
    120,
    121,
    122,
    123,
    124,
    125,
    126,
    127,
    128,
    129,
    130,
    131,
    132,
    133,
    134,
    135,
    136,
    137,
    138,
    139,
    140,
    141,
    142,
    143,
    144,
    145,
    146,
    147,
    148,
    149,
    150,
    151,
    152,
    153,
    154,
    155,
    156,
    157,
    158,
    159,
    160,
    161,
    162,
    163,
    164,
    165,
    166,
    167,
    168,
    169,
    170,
    171,
    172,
    173,
    174,
    175,
    176,
    177,
    178,
    179,
    180,
    181,
    182,
    183,
    184,
    185,
    186,
    187,
    188,
    189,
    190,
    191,
    192,
    193,
    194,
    195,
    196,
    197,
    198,
    199,
    200,
    201,
    202,
    203,
    204,
    205,
    206,
    207,
    208,
    209,
    210,
    211,
    212,
    213,
    214,
    215,
    216,
    217,
    218,
    219,
    220,
    221,
    222,
    223,
    224,
    225,
    226,
    227,
    228,
    229,
    230,
    231,
    232,
    233,
    234,
    235,
    236,
    237,
    238,
    239,
    240,
    241,
    242,
    243,
    244,
    245,
    246,
    247,
    248,
    249,
    250,
    251,
    252,
    253,
    254,
    255
  ]
};

// src/cmaps/actc.json
var actc_default = {
  R: [0, 0, 24, 248, 255],
  G: [0, 0, 177, 254, 0],
  B: [0, 136, 0, 0, 0],
  A: [0, 32, 64, 78, 128],
  I: [0, 64, 128, 156, 255]
};

// src/cmaps/batlow.json
var batlow_default = {
  R: [1, 10, 15, 17, 21, 27, 36, 49, 65, 82, 99, 118, 140, 161, 183, 203, 222, 238, 248, 253, 253, 253, 252, 250],
  G: [25, 42, 56, 68, 79, 88, 97, 105, 111, 116, 122, 127, 133, 138, 142, 146, 150, 155, 162, 170, 178, 186, 195, 204],
  B: [89, 92, 95, 96, 98, 98, 96, 91, 83, 74, 64, 55, 46, 43, 49, 62, 79, 100, 126, 151, 175, 199, 223, 250],
  A: [0, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64],
  I: [0, 11, 22, 33, 44, 55, 66, 77, 88, 99, 110, 121, 133, 144, 155, 166, 177, 188, 199, 210, 221, 232, 243, 255]
};

// src/cmaps/bcgwhw_dark.json
var bcgwhw_dark_default = {
  R: [
    0,
    248,
    242,
    235,
    229,
    222,
    212,
    202,
    193,
    183,
    173,
    145,
    117,
    78,
    39,
    0,
    0,
    0,
    0,
    0,
    0,
    64,
    128,
    191,
    255,
    255,
    255,
    255,
    255,
    238,
    221,
    204,
    214,
    224,
    235,
    245,
    255,
    249,
    242,
    236,
    229,
    223,
    217,
    210,
    204,
    197,
    191,
    184,
    176,
    169,
    161,
    154,
    146,
    139,
    132,
    124,
    117,
    109,
    102,
    0
  ],
  G: [
    0,
    251,
    247,
    243,
    239,
    235,
    235,
    235,
    235,
    235,
    235,
    216,
    197,
    174,
    150,
    127,
    153,
    178,
    204,
    229,
    255,
    255,
    255,
    255,
    255,
    220,
    185,
    150,
    115,
    77,
    38,
    0,
    31,
    61,
    92,
    122,
    153,
    138,
    122,
    107,
    92,
    77,
    61,
    46,
    31,
    15,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0
  ],
  B: [
    0,
    254,
    253,
    252,
    251,
    250,
    251,
    252,
    253,
    254,
    255,
    226,
    198,
    132,
    66,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    40,
    80,
    120,
    160,
    200,
    199,
    198,
    197,
    196,
    196,
    195,
    194,
    193,
    192,
    191,
    184,
    176,
    169,
    161,
    154,
    146,
    139,
    132,
    124,
    117,
    109,
    102,
    0
  ],
  A: [
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256
  ],
  I: [
    0,
    4,
    8,
    12,
    16,
    20,
    24,
    28,
    32,
    36,
    40,
    44,
    48,
    52,
    56,
    60,
    64,
    68,
    72,
    76,
    80,
    84,
    88,
    92,
    96,
    100,
    104,
    108,
    112,
    116,
    120,
    124,
    128,
    132,
    136,
    140,
    144,
    148,
    152,
    156,
    160,
    164,
    168,
    172,
    176,
    180,
    184,
    188,
    192,
    196,
    200,
    204,
    208,
    212,
    216,
    220,
    224,
    228,
    232,
    236,
    240,
    244,
    248,
    255
  ]
};

// src/cmaps/bcgwhw.json
var bcgwhw_default = {
  R: [
    255,
    248,
    242,
    235,
    229,
    222,
    212,
    202,
    193,
    183,
    173,
    145,
    117,
    78,
    39,
    0,
    0,
    0,
    0,
    0,
    0,
    64,
    128,
    191,
    255,
    255,
    255,
    255,
    255,
    238,
    221,
    204,
    214,
    224,
    235,
    245,
    255,
    249,
    242,
    236,
    229,
    223,
    217,
    210,
    204,
    197,
    191,
    184,
    176,
    169,
    161,
    154,
    146,
    139,
    132,
    124,
    117,
    109,
    102,
    0
  ],
  G: [
    255,
    251,
    247,
    243,
    239,
    235,
    235,
    235,
    235,
    235,
    235,
    216,
    197,
    174,
    150,
    127,
    153,
    178,
    204,
    229,
    255,
    255,
    255,
    255,
    255,
    220,
    185,
    150,
    115,
    77,
    38,
    0,
    31,
    61,
    92,
    122,
    153,
    138,
    122,
    107,
    92,
    77,
    61,
    46,
    31,
    15,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0
  ],
  B: [
    255,
    254,
    253,
    252,
    251,
    250,
    251,
    252,
    253,
    254,
    255,
    226,
    198,
    132,
    66,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    40,
    80,
    120,
    160,
    200,
    199,
    198,
    197,
    196,
    196,
    195,
    194,
    193,
    192,
    191,
    184,
    176,
    169,
    161,
    154,
    146,
    139,
    132,
    124,
    117,
    109,
    102,
    0
  ],
  A: [
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256,
    256
  ],
  I: [
    0,
    4,
    8,
    12,
    16,
    20,
    24,
    28,
    32,
    36,
    40,
    44,
    48,
    52,
    56,
    60,
    64,
    68,
    72,
    76,
    80,
    84,
    88,
    92,
    96,
    100,
    104,
    108,
    112,
    116,
    120,
    124,
    128,
    132,
    136,
    140,
    144,
    148,
    152,
    156,
    160,
    164,
    168,
    172,
    176,
    180,
    184,
    188,
    192,
    196,
    200,
    204,
    208,
    212,
    216,
    220,
    224,
    228,
    232,
    236,
    240,
    244,
    248,
    255
  ]
};

// src/cmaps/blue2red.json
var blue2red_default = {
  R: [0, 0, 0, 0, 196, 255],
  G: [0, 32, 128, 128, 128, 32],
  B: [0, 255, 196, 0, 0, 0],
  A: [0, 128, 64, 64, 64, 128],
  I: [0, 1, 64, 128, 192, 255]
};

// src/cmaps/blue2magenta.json
var blue2magenta_default = {
  R: [0, 255],
  G: [0, 0],
  B: [255, 255],
  A: [0, 128],
  I: [0, 255]
};

// src/cmaps/blue2cyan.json
var blue2cyan_default = {
  R: [0, 0],
  G: [0, 255],
  B: [255, 255],
  A: [0, 128],
  I: [0, 255]
};

// src/cmaps/bluegrn.json
var bluegrn_default = {
  R: [0, 0, 0, 0],
  G: [0, 1, 128, 255],
  B: [0, 222, 127, 32],
  A: [0, 0, 64, 128],
  I: [0, 1, 128, 255]
};

// src/cmaps/blue.json
var blue_default = {
  R: [0, 0, 0],
  G: [0, 0, 0],
  B: [0, 128, 255],
  A: [0, 64, 128],
  I: [0, 128, 255]
};

// src/cmaps/bone.json
var bone_default = {
  R: [0, 103, 255],
  G: [0, 126, 255],
  B: [0, 165, 255],
  A: [0, 76, 128],
  I: [0, 153, 255]
};

// src/cmaps/bronze.json
var bronze_default = {
  R: [0, 43, 103, 199, 216, 255],
  G: [0, 0, 37, 155, 213, 255],
  B: [0, 0, 20, 97, 201, 255],
  A: [0, 44, 48, 54, 56, 56],
  I: [0, 64, 128, 196, 240, 255]
};

// src/cmaps/cet_l17.json
var cet_l17_default = {
  R: [
    0,
    9,
    24,
    33,
    40,
    46,
    52,
    57,
    62,
    66,
    70,
    74,
    78,
    81,
    85,
    88,
    91,
    94,
    98,
    101,
    103,
    106,
    109,
    112,
    114,
    117,
    119,
    121,
    124,
    126,
    128,
    130,
    133,
    135,
    137,
    139,
    141,
    143,
    145,
    147,
    149,
    151,
    153,
    155,
    157,
    159,
    161,
    163,
    164,
    166,
    168,
    169,
    171,
    172,
    174,
    175,
    177,
    178,
    180,
    181,
    183,
    184,
    186,
    187,
    189,
    190,
    191,
    193,
    194,
    196,
    197,
    198,
    199,
    200,
    201,
    202,
    203,
    204,
    205,
    206,
    207,
    208,
    209,
    210,
    211,
    212,
    213,
    214,
    215,
    216,
    217,
    218,
    219,
    220,
    221,
    222,
    222,
    223,
    224,
    224,
    225,
    226,
    226,
    227,
    228,
    228,
    229,
    229,
    230,
    231,
    231,
    232,
    233,
    233,
    234,
    234,
    235,
    235,
    236,
    236,
    236,
    237,
    237,
    237,
    238,
    238,
    238,
    239,
    239,
    239,
    240,
    240,
    240,
    241,
    241,
    241,
    242,
    242,
    242,
    243,
    243,
    243,
    243,
    243,
    243,
    243,
    243,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    245,
    245,
    245,
    245,
    245,
    245,
    245,
    245,
    245,
    245,
    245,
    245,
    245,
    245,
    245,
    245,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    244,
    243,
    243,
    243,
    243,
    243,
    243,
    243,
    243,
    243,
    243,
    243,
    243,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    241,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    242,
    243,
    243,
    244,
    244,
    245,
    246,
    246,
    247,
    247,
    248,
    249,
    249,
    250,
    250,
    251,
    251,
    252,
    252,
    253,
    253,
    254,
    254,
    254
  ],
  G: [
    42,
    41,
    41,
    41,
    40,
    40,
    40,
    39,
    39,
    39,
    38,
    38,
    37,
    37,
    36,
    36,
    35,
    35,
    34,
    34,
    33,
    32,
    32,
    31,
    30,
    30,
    30,
    30,
    29,
    29,
    29,
    29,
    28,
    28,
    28,
    27,
    27,
    27,
    26,
    26,
    26,
    25,
    25,
    25,
    24,
    24,
    24,
    24,
    24,
    25,
    25,
    26,
    27,
    27,
    28,
    28,
    29,
    30,
    30,
    31,
    31,
    32,
    33,
    33,
    34,
    34,
    35,
    36,
    36,
    37,
    38,
    39,
    40,
    42,
    43,
    44,
    45,
    47,
    48,
    49,
    50,
    51,
    53,
    54,
    55,
    56,
    57,
    58,
    59,
    60,
    62,
    63,
    64,
    65,
    66,
    68,
    69,
    71,
    72,
    73,
    75,
    76,
    77,
    79,
    80,
    81,
    83,
    84,
    85,
    87,
    88,
    89,
    90,
    92,
    93,
    94,
    95,
    97,
    98,
    100,
    101,
    102,
    104,
    105,
    107,
    108,
    109,
    111,
    112,
    113,
    115,
    116,
    117,
    119,
    120,
    121,
    122,
    124,
    125,
    126,
    128,
    129,
    130,
    132,
    133,
    134,
    136,
    137,
    138,
    140,
    141,
    142,
    144,
    145,
    146,
    147,
    149,
    150,
    151,
    153,
    154,
    155,
    156,
    158,
    159,
    160,
    161,
    163,
    164,
    165,
    167,
    168,
    169,
    170,
    172,
    173,
    174,
    175,
    176,
    178,
    179,
    180,
    181,
    183,
    184,
    185,
    186,
    187,
    189,
    190,
    191,
    192,
    193,
    194,
    196,
    197,
    198,
    199,
    200,
    201,
    203,
    204,
    205,
    206,
    207,
    208,
    209,
    211,
    212,
    213,
    214,
    215,
    216,
    217,
    218,
    219,
    220,
    221,
    222,
    223,
    224,
    225,
    226,
    228,
    229,
    230,
    231,
    232,
    233,
    234,
    235,
    236,
    237,
    238,
    238,
    239,
    240,
    241,
    241,
    242,
    243,
    244,
    244,
    245,
    246,
    247,
    248,
    248,
    249,
    250,
    251,
    251,
    252,
    253,
    254,
    255
  ],
  B: [
    167,
    167,
    166,
    166,
    166,
    165,
    165,
    165,
    164,
    164,
    164,
    163,
    163,
    162,
    162,
    162,
    161,
    161,
    161,
    160,
    160,
    160,
    159,
    159,
    158,
    158,
    157,
    157,
    156,
    155,
    155,
    154,
    154,
    153,
    153,
    152,
    151,
    151,
    150,
    150,
    149,
    149,
    148,
    147,
    147,
    146,
    146,
    145,
    144,
    144,
    143,
    143,
    142,
    141,
    141,
    140,
    139,
    139,
    138,
    137,
    137,
    136,
    135,
    135,
    134,
    133,
    133,
    132,
    131,
    131,
    130,
    130,
    129,
    128,
    128,
    127,
    127,
    126,
    125,
    125,
    124,
    124,
    123,
    123,
    122,
    121,
    121,
    120,
    119,
    119,
    118,
    118,
    117,
    116,
    116,
    116,
    115,
    115,
    114,
    114,
    113,
    113,
    113,
    112,
    112,
    111,
    111,
    110,
    110,
    109,
    109,
    109,
    108,
    108,
    107,
    107,
    106,
    106,
    106,
    106,
    106,
    105,
    105,
    105,
    105,
    105,
    105,
    105,
    104,
    104,
    104,
    104,
    104,
    103,
    103,
    103,
    103,
    103,
    102,
    102,
    102,
    103,
    103,
    103,
    103,
    104,
    104,
    104,
    104,
    104,
    105,
    105,
    105,
    105,
    106,
    106,
    106,
    106,
    106,
    107,
    107,
    107,
    107,
    108,
    108,
    109,
    110,
    111,
    111,
    112,
    113,
    113,
    114,
    115,
    115,
    116,
    117,
    117,
    118,
    119,
    120,
    120,
    121,
    122,
    122,
    123,
    124,
    125,
    126,
    128,
    129,
    130,
    131,
    132,
    134,
    135,
    136,
    137,
    138,
    140,
    141,
    142,
    143,
    144,
    146,
    147,
    148,
    149,
    150,
    152,
    153,
    155,
    157,
    158,
    160,
    162,
    163,
    165,
    167,
    168,
    170,
    172,
    173,
    175,
    177,
    178,
    180,
    182,
    183,
    185,
    187,
    188,
    190,
    193,
    196,
    199,
    201,
    204,
    207,
    210,
    212,
    215,
    218,
    221,
    224,
    226,
    229,
    232,
    235,
    238,
    240,
    243,
    246,
    249,
    252,
    255
  ],
  A: [
    0,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64
  ],
  I: [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30,
    31,
    32,
    33,
    34,
    35,
    36,
    37,
    38,
    39,
    40,
    41,
    42,
    43,
    44,
    45,
    46,
    47,
    48,
    49,
    50,
    51,
    52,
    53,
    54,
    55,
    56,
    57,
    58,
    59,
    60,
    61,
    62,
    63,
    64,
    65,
    66,
    67,
    68,
    69,
    70,
    71,
    72,
    73,
    74,
    75,
    76,
    77,
    78,
    79,
    80,
    81,
    82,
    83,
    84,
    85,
    86,
    87,
    88,
    89,
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
    99,
    100,
    101,
    102,
    103,
    104,
    105,
    106,
    107,
    108,
    109,
    110,
    111,
    112,
    113,
    114,
    115,
    116,
    117,
    118,
    119,
    120,
    121,
    122,
    123,
    124,
    125,
    126,
    127,
    128,
    129,
    130,
    131,
    132,
    133,
    134,
    135,
    136,
    137,
    138,
    139,
    140,
    141,
    142,
    143,
    144,
    145,
    146,
    147,
    148,
    149,
    150,
    151,
    152,
    153,
    154,
    155,
    156,
    157,
    158,
    159,
    160,
    161,
    162,
    163,
    164,
    165,
    166,
    167,
    168,
    169,
    170,
    171,
    172,
    173,
    174,
    175,
    176,
    177,
    178,
    179,
    180,
    181,
    182,
    183,
    184,
    185,
    186,
    187,
    188,
    189,
    190,
    191,
    192,
    193,
    194,
    195,
    196,
    197,
    198,
    199,
    200,
    201,
    202,
    203,
    204,
    205,
    206,
    207,
    208,
    209,
    210,
    211,
    212,
    213,
    214,
    215,
    216,
    217,
    218,
    219,
    220,
    221,
    222,
    223,
    224,
    225,
    226,
    227,
    228,
    229,
    230,
    231,
    232,
    233,
    234,
    235,
    236,
    237,
    238,
    239,
    240,
    241,
    242,
    243,
    244,
    245,
    246,
    247,
    248,
    249,
    250,
    251,
    252,
    253,
    254,
    255
  ]
};

// src/cmaps/cividis.json
var cividis_default = {
  R: [0, 86, 166, 255],
  G: [32, 92, 156, 233],
  B: [76, 108, 117, 69],
  A: [0, 56, 80, 88],
  I: [0, 64, 192, 255]
};

// src/cmaps/cool.json
var cool_default = {
  R: [0, 0, 0],
  G: [127, 196, 254],
  B: [255, 255, 255],
  A: [0, 64, 128],
  I: [0, 128, 255]
};

// src/cmaps/copper2.json
var copper2_default = {
  R: [0, 61, 122, 183, 244, 255],
  G: [0, 41, 81, 122, 163, 255],
  B: [0, 25, 51, 76, 102, 255],
  A: [0, 25, 51, 71, 102, 128],
  I: [0, 51, 102, 153, 204, 255]
};

// src/cmaps/copper.json
var copper_default = {
  R: [0, 61, 122, 183, 244, 255],
  G: [0, 41, 81, 122, 163, 203],
  B: [0, 25, 51, 76, 102, 127],
  A: [0, 25, 51, 71, 102, 128],
  I: [0, 51, 102, 153, 204, 255]
};

// src/cmaps/ct_airways.json
var ct_airways_default = {
  min: -643,
  max: -235,
  R: [0, 0, 0],
  G: [154, 154, 154],
  B: [179, 179, 101],
  A: [0, 32, 0],
  I: [0, 163, 255]
};

// src/cmaps/ct_artery.json
var ct_artery_default = {
  min: 114,
  max: 800,
  R: [0, 255, 255, 255],
  G: [0, 0, 129, 255],
  B: [0, 0, 0, 255],
  A: [0, 64, 88, 228],
  I: [0, 80, 160, 255]
};

// src/cmaps/ct_bones.json
var ct_bones_default = {
  min: 180,
  max: 600,
  R: [0, 0, 113, 255],
  G: [0, 0, 109, 250],
  B: [0, 0, 101, 245],
  A: [0, 0, 100, 160],
  I: [0, 1, 128, 255]
};

// src/cmaps/ct_brain_gray.json
var ct_brain_gray_default = {
  min: -10,
  max: 110,
  R: [0, 127, 255],
  G: [0, 127, 255],
  B: [0, 127, 255],
  A: [0, 48, 128],
  I: [0, 124, 255]
};

// src/cmaps/ct_brain.json
var ct_brain_default = {
  min: -10,
  max: 110,
  R: [0, 199, 255],
  G: [0, 127, 255],
  B: [0, 127, 255],
  A: [0, 48, 128],
  I: [0, 124, 255]
};

// src/cmaps/ct_cardiac.json
var ct_cardiac_default = {
  min: -80,
  max: 1e3,
  R: [0, 189, 150, 150, 150, 150, 255],
  G: [0, 169, 54, 54, 54, 54, 240],
  B: [0, 153, 52, 52, 52, 52, 242],
  A: [0, 32, 64, 0, 0, 64, 64],
  I: [0, 1, 82, 92, 234, 242, 255]
};

// src/cmaps/ct_head.json
var ct_head_default = {
  min: -590,
  max: 600,
  R: [0, 241, 241, 248, 248, 178, 178, 232, 255, 255, 255],
  G: [0, 156, 156, 222, 222, 36, 36, 51, 255, 255, 255],
  B: [0, 130, 130, 169, 169, 24, 24, 37, 255, 255, 255],
  A: [0, 8, 0, 0, 0, 64, 64, 0, 0, 222, 222],
  I: [0, 2, 3, 64, 122, 142, 172, 182, 252, 253, 255]
};

// src/cmaps/ct_kidneys.json
var ct_kidneys_default = {
  min: 114,
  max: 302,
  R: [0, 255, 255],
  G: [0, 129, 255],
  B: [0, 0, 255],
  A: [0, 88, 228],
  I: [0, 103, 255]
};

// src/cmaps/ct_liver.json
var ct_liver_default = {
  min: -23,
  max: 246,
  R: [0, 44, 255, 255, 255],
  G: [0, 128, 90, 255, 255],
  B: [0, 0, 70, 0, 255],
  A: [0, 0, 82, 184, 228],
  I: [0, 64, 131, 196, 255]
};

// src/cmaps/ct_muscles.json
var ct_muscles_default = {
  min: -100,
  max: 246,
  R: [0, 128, 159, 255, 255, 255, 255],
  G: [0, 0, 56, 90, 0, 255, 255],
  B: [0, 0, 41, 70, 0, 0, 255],
  A: [0, 63, 105, 135, 167, 184, 228],
  I: [0, 100, 128, 155, 180, 209, 255]
};

// src/cmaps/ct_scalp.json
var ct_scalp_default = {
  min: -590,
  max: 600,
  R: [0, 241, 241, 248, 248, 178, 232, 255, 255],
  G: [0, 156, 156, 222, 222, 36, 51, 255, 255],
  B: [0, 130, 130, 169, 169, 24, 37, 255, 255],
  A: [0, 63, 105, 135, 167, 184, 228, 228, 228],
  I: [0, 1, 52, 127, 137, 162, 172, 252, 255]
};

// src/cmaps/ct_skull.json
var ct_skull_default = {
  min: 140,
  max: 1024,
  R: [0, 2, 113, 255],
  G: [0, 1, 109, 250],
  B: [0, 1, 101, 245],
  A: [0, 1, 96, 168],
  I: [0, 1, 128, 255]
};

// src/cmaps/ct_soft.json
var ct_soft_default = {
  min: -923,
  max: 679,
  R: [0, 0, 0, 0, 0, 255, 255, 255],
  G: [154, 154, 154, 154, 0, 0, 254, 255],
  B: [179, 179, 179, 179, 0, 0, 0, 255],
  A: [0, 3, 8, 0, 0, 10, 15, 20],
  I: [0, 30, 62, 88, 170, 200, 232, 255]
};

// src/cmaps/ct_soft_tissue.json
var ct_soft_tissue_default = {
  min: -10,
  max: 110,
  R: [0, 199, 255],
  G: [0, 127, 255],
  B: [0, 127, 255],
  A: [0, 48, 128],
  I: [0, 124, 255]
};

// src/cmaps/ct_surface.json
var ct_surface_default = {
  min: -600,
  max: 100,
  R: [0, 134, 255],
  G: [0, 109, 250],
  B: [0, 101, 245],
  A: [0, 60, 148],
  I: [0, 128, 255]
};

// src/cmaps/ct_vessels.json
var ct_vessels_default = {
  min: 114,
  max: 246,
  R: [0, 255, 255],
  G: [0, 128, 255],
  B: [0, 128, 255],
  A: [0, 64, 96],
  I: [0, 87, 255]
};

// src/cmaps/ct_w_contrast.json
var ct_w_contrast_default = {
  min: 50,
  max: 1e3,
  R: [98, 210, 169, 128, 255],
  G: [94, 26, 77, 128, 255],
  B: [45, 21, 74, 128, 255],
  A: [0, 25, 0, 4, 168],
  I: [0, 41, 87, 154, 255]
};

// src/cmaps/cubehelix.json
var cubehelix_default = {
  R: [
    0,
    13,
    21,
    26,
    27,
    25,
    22,
    21,
    22,
    28,
    39,
    54,
    75,
    98,
    124,
    148,
    171,
    189,
    202,
    210,
    213,
    211,
    206,
    200,
    195,
    193,
    195,
    201,
    211,
    225,
    240,
    255
  ],
  G: [
    0,
    5,
    11,
    20,
    31,
    44,
    58,
    72,
    86,
    99,
    109,
    116,
    120,
    122,
    122,
    122,
    121,
    121,
    124,
    129,
    137,
    147,
    161,
    175,
    190,
    205,
    218,
    229,
    238,
    245,
    251,
    255
  ],
  B: [
    0,
    14,
    30,
    46,
    61,
    71,
    77,
    78,
    75,
    68,
    60,
    52,
    48,
    47,
    53,
    65,
    83,
    105,
    131,
    157,
    183,
    205,
    222,
    235,
    241,
    243,
    242,
    240,
    239,
    240,
    245,
    255
  ],
  A: [
    0,
    4,
    8,
    12,
    17,
    21,
    25,
    29,
    33,
    37,
    41,
    45,
    50,
    54,
    58,
    62,
    66,
    70,
    74,
    78,
    83,
    87,
    91,
    95,
    99,
    103,
    107,
    111,
    116,
    120,
    124,
    128
  ],
  I: [
    0,
    8,
    16,
    25,
    33,
    41,
    49,
    58,
    66,
    74,
    82,
    90,
    99,
    107,
    115,
    123,
    132,
    140,
    148,
    156,
    165,
    173,
    181,
    189,
    197,
    206,
    214,
    222,
    230,
    239,
    247,
    255
  ]
};

// src/cmaps/electric_blue.json
var electric_blue_default = {
  R: [0, 10, 136, 255],
  G: [0, 39, 220, 255],
  B: [0, 223, 253, 255],
  A: [0, 48, 64, 70],
  I: [0, 92, 192, 255]
};

// src/cmaps/freesurfer.json
var freesurfer_default = {
  R: [
    0,
    245,
    205,
    120,
    196,
    220,
    230,
    0,
    122,
    236,
    12,
    204,
    42,
    119,
    220,
    103,
    60,
    255,
    165,
    160,
    0,
    245,
    205,
    120,
    196,
    220,
    230,
    0,
    122,
    236,
    13,
    220,
    103,
    255,
    165,
    160,
    0,
    120,
    200,
    255,
    255,
    164,
    164,
    164,
    234,
    0,
    0,
    0,
    0,
    0
  ],
  G: [
    0,
    245,
    62,
    18,
    58,
    248,
    148,
    118,
    186,
    13,
    48,
    182,
    204,
    159,
    216,
    255,
    60,
    165,
    42,
    32,
    200,
    245,
    62,
    18,
    58,
    248,
    148,
    118,
    186,
    13,
    48,
    216,
    255,
    165,
    42,
    32,
    200,
    190,
    70,
    148,
    148,
    108,
    108,
    108,
    169,
    0,
    0,
    0,
    0,
    0
  ],
  B: [
    0,
    245,
    78,
    134,
    250,
    164,
    34,
    14,
    220,
    176,
    255,
    142,
    164,
    176,
    20,
    255,
    60,
    0,
    42,
    240,
    200,
    245,
    78,
    134,
    250,
    164,
    34,
    14,
    220,
    176,
    255,
    20,
    255,
    0,
    42,
    240,
    221,
    150,
    255,
    10,
    10,
    226,
    226,
    226,
    30,
    64,
    112,
    160,
    208,
    255
  ],
  A: [
    0,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64
  ],
  I: [
    0,
    2,
    3,
    4,
    5,
    7,
    8,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    24,
    26,
    28,
    30,
    31,
    41,
    42,
    43,
    44,
    46,
    47,
    49,
    50,
    51,
    52,
    53,
    54,
    58,
    60,
    62,
    63,
    72,
    77,
    78,
    79,
    80,
    81,
    82,
    85,
    251,
    252,
    253,
    254,
    255
  ]
};

// src/cmaps/ge_color.json
var ge_color_default = {
  R: [0, 0, 128, 255, 255],
  G: [0, 128, 0, 128, 255],
  B: [0, 125, 255, 0, 255],
  A: [0, 32, 64, 96, 128],
  I: [0, 63, 128, 192, 255]
};

// src/cmaps/gold.json
var gold_default = {
  R: [0, 142, 227, 255],
  G: [0, 85, 170, 255],
  B: [0, 14, 76, 255],
  A: [0, 42, 84, 128],
  I: [0, 85, 170, 255]
};

// src/cmaps/gray.json
var gray_default = {
  R: [0, 255],
  G: [0, 255],
  B: [0, 255],
  A: [0, 128],
  I: [0, 255]
};

// src/cmaps/green.json
var green_default = {
  R: [0, 0, 0],
  G: [0, 128, 255],
  B: [0, 0, 0],
  A: [0, 64, 128],
  I: [0, 128, 255]
};

// src/cmaps/green2cyan.json
var green2cyan_default = {
  R: [0, 0, 0],
  G: [72, 72, 255],
  B: [0, 255, 255],
  A: [0, 64, 128],
  I: [0, 88, 255]
};

// src/cmaps/green2orange.json
var green2orange_default = {
  R: [0, 255, 255],
  G: [72, 88, 255],
  B: [0, 0, 0],
  A: [0, 64, 128],
  I: [0, 88, 255]
};

// src/cmaps/hotiron.json
var hotiron_default = {
  R: [0, 255, 255, 255],
  G: [0, 0, 126, 255],
  B: [0, 0, 0, 255],
  A: [0, 64, 96, 128],
  I: [0, 128, 191, 255]
};

// src/cmaps/hot.json
var hot_default = {
  R: [3, 255, 255, 255],
  G: [0, 0, 255, 255],
  B: [0, 0, 0, 255],
  A: [0, 48, 96, 128],
  I: [0, 95, 191, 255]
};

// src/cmaps/hsv.json
var hsv_default = {
  R: [255, 255, 0, 0, 0, 255, 255],
  G: [0, 255, 255, 255, 0, 0, 0],
  B: [0, 0, 0, 255, 255, 255, 0],
  A: [0, 14, 28, 43, 57, 71, 85],
  I: [0, 43, 85, 128, 170, 213, 255]
};

// src/cmaps/inferno.json
var inferno_default = {
  R: [0, 120, 237, 240],
  G: [0, 28, 105, 249],
  B: [4, 109, 37, 33],
  A: [0, 56, 80, 88],
  I: [0, 64, 192, 255]
};

// src/cmaps/jet.json
var jet_default = {
  R: [0, 0, 127, 255, 127],
  G: [0, 127, 255, 127, 0],
  B: [127, 255, 127, 0, 0],
  A: [0, 32, 64, 96, 128],
  I: [0, 63, 128, 192, 255]
};

// src/cmaps/kry.json
var kry_default = {
  R: [0, 255, 255],
  G: [0, 0, 255],
  B: [0, 0, 0],
  A: [0, 64, 64],
  I: [0, 86, 255]
};

// src/cmaps/linspecer.json
var linspecer_default = {
  R: [94, 50, 90, 152, 215, 238, 249, 254, 252, 241, 209, 158],
  G: [79, 131, 186, 214, 240, 244, 237, 210, 157, 100, 57, 1],
  B: [162, 189, 167, 164, 155, 169, 168, 123, 86, 68, 79, 66],
  A: [0, 12, 23, 35, 47, 58, 70, 81, 93, 105, 116, 128],
  I: [0, 23, 46, 70, 93, 116, 139, 162, 185, 209, 232, 255]
};

// src/cmaps/lipari.json
var lipari_default = {
  R: [3, 7, 13, 25, 45, 67, 84, 98, 110, 124, 138, 154, 173, 191, 210, 225, 233, 233, 231, 229, 230, 235, 243, 253],
  G: [19, 33, 48, 63, 77, 87, 92, 94, 95, 96, 96, 97, 99, 101, 106, 118, 133, 149, 163, 177, 192, 208, 225, 245],
  B: [38, 58, 79, 99, 114, 122, 122, 120, 117, 113, 110, 105, 101, 97, 94, 96, 103, 112, 122, 134, 149, 169, 192, 218],
  A: [0, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64],
  I: [0, 11, 22, 33, 44, 55, 66, 77, 88, 99, 110, 121, 133, 144, 155, 166, 177, 188, 199, 210, 221, 232, 243, 255]
};

// src/cmaps/magma.json
var magma_default = {
  R: [0, 148, 183, 223, 247, 252],
  G: [0, 44, 55, 74, 112, 253],
  B: [4, 128, 121, 104, 92, 191],
  A: [0, 44, 53, 64, 75, 107],
  I: [0, 107, 128, 154, 179, 255]
};

// src/cmaps/mako.json
var mako_default = {
  R: [11, 59, 55, 222],
  G: [4, 45, 165, 245],
  B: [5, 91, 172, 229],
  A: [0, 23, 70, 107],
  I: [0, 56, 167, 255]
};

// src/cmaps/navia.json
var navia_default = {
  R: [3, 5, 6, 8, 12, 19, 28, 36, 42, 48, 54, 61, 68, 76, 86, 96, 110, 129, 154, 181, 205, 224, 239, 252],
  G: [19, 32, 44, 58, 72, 86, 98, 108, 116, 122, 128, 134, 141, 149, 157, 167, 178, 191, 204, 216, 225, 232, 239, 244],
  B: [39, 60, 82, 104, 123, 137, 144, 145, 142, 138, 134, 130, 126, 121, 116, 110, 105, 104, 112, 131, 155, 178, 198, 217],
  A: [0, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64],
  I: [0, 11, 22, 33, 44, 55, 66, 77, 88, 99, 110, 121, 133, 144, 155, 166, 177, 188, 199, 210, 221, 232, 243, 255]
};

// src/cmaps/nih.json
var nih_default = {
  R: [0, 85, 0, 0, 0, 0, 0, 0, 85, 255, 255, 255, 172],
  G: [0, 0, 0, 0, 85, 170, 255, 255, 255, 255, 85, 0, 0],
  B: [0, 170, 85, 255, 255, 170, 170, 0, 85, 0, 0, 0, 0],
  A: [0, 5, 10, 21, 26, 32, 37, 42, 48, 53, 64, 72, 85],
  I: [0, 15, 31, 63, 79, 95, 111, 127, 143, 159, 191, 217, 255]
};

// src/cmaps/plasma.json
var plasma_default = {
  R: [13, 156, 237, 240],
  G: [8, 23, 121, 249],
  B: [135, 158, 83, 33],
  A: [0, 56, 80, 88],
  I: [0, 64, 192, 255]
};

// src/cmaps/random.json
var random_default = {
  R: [
    208,
    71,
    33,
    192,
    32,
    195,
    208,
    173,
    233,
    202,
    25,
    210,
    145,
    89,
    87,
    245,
    246,
    38,
    3,
    25,
    57,
    167,
    245,
    86,
    227,
    208,
    81,
    64,
    90,
    199,
    140,
    48,
    212,
    180,
    70,
    120,
    9,
    192,
    245,
    177,
    65,
    157,
    9,
    193,
    100,
    181,
    125,
    145,
    62,
    8,
    108,
    36,
    140,
    237,
    242,
    248,
    161,
    189,
    41,
    114,
    65,
    121,
    97,
    50,
    238,
    149,
    44,
    214,
    124,
    167,
    40,
    167,
    127,
    178,
    231,
    30,
    173,
    244,
    193,
    203,
    204,
    238,
    139,
    135,
    71,
    234,
    234,
    217,
    66,
    14,
    129,
    19,
    97,
    165,
    112,
    244,
    35,
    73,
    192,
    12,
    149,
    71,
    33,
    192,
    32,
    195,
    208,
    173,
    233,
    202,
    25,
    210,
    145,
    89,
    87,
    245,
    246,
    38,
    3,
    25,
    57,
    167,
    245,
    86,
    227,
    208,
    81,
    64,
    90,
    199,
    140,
    48,
    212,
    180,
    70,
    120,
    9,
    192,
    245,
    177,
    65,
    157,
    9,
    193,
    100,
    181,
    125,
    145,
    62,
    8,
    108,
    36,
    140,
    237,
    242,
    248,
    161,
    189,
    41,
    114,
    65,
    121,
    97,
    50,
    238,
    149,
    44,
    214,
    124,
    167,
    40,
    167,
    127,
    178,
    231,
    30,
    173,
    244,
    193,
    203,
    204,
    238,
    139,
    135,
    71,
    234,
    234,
    217,
    66,
    14,
    129,
    19,
    97,
    165,
    112,
    244,
    35,
    73,
    192,
    12,
    149,
    71,
    33,
    192,
    32,
    195,
    208,
    173,
    233,
    202,
    25,
    210,
    145,
    89,
    87,
    245,
    246,
    38,
    3,
    25,
    57,
    167,
    245,
    86,
    227,
    208,
    81,
    64,
    90,
    199,
    140,
    48,
    212,
    180,
    70,
    120,
    9,
    192,
    245,
    177,
    65,
    157,
    9,
    193,
    100,
    181,
    125,
    145,
    62,
    8,
    108,
    36,
    140,
    237,
    242,
    248
  ],
  G: [
    182,
    46,
    78,
    199,
    79,
    89,
    41,
    208,
    135,
    20,
    154,
    35,
    21,
    43,
    230,
    113,
    191,
    147,
    208,
    37,
    28,
    27,
    86,
    203,
    25,
    209,
    148,
    187,
    139,
    111,
    48,
    102,
    76,
    110,
    106,
    130,
    37,
    160,
    34,
    222,
    90,
    165,
    245,
    222,
    102,
    47,
    19,
    130,
    4,
    232,
    137,
    211,
    240,
    11,
    140,
    21,
    42,
    22,
    241,
    61,
    99,
    115,
    199,
    166,
    114,
    190,
    204,
    60,
    233,
    66,
    115,
    230,
    125,
    103,
    203,
    125,
    13,
    176,
    94,
    131,
    39,
    198,
    167,
    124,
    67,
    175,
    254,
    1,
    15,
    198,
    62,
    237,
    159,
    31,
    218,
    58,
    244,
    47,
    61,
    67,
    94,
    46,
    78,
    199,
    79,
    89,
    41,
    208,
    135,
    20,
    154,
    35,
    21,
    43,
    230,
    113,
    191,
    147,
    208,
    37,
    28,
    27,
    86,
    203,
    25,
    209,
    148,
    187,
    139,
    111,
    48,
    102,
    76,
    110,
    106,
    130,
    37,
    160,
    34,
    222,
    90,
    165,
    245,
    222,
    102,
    47,
    19,
    130,
    4,
    232,
    137,
    211,
    240,
    11,
    140,
    21,
    42,
    22,
    241,
    61,
    99,
    115,
    199,
    166,
    114,
    190,
    204,
    60,
    233,
    66,
    115,
    230,
    125,
    103,
    203,
    125,
    13,
    176,
    94,
    131,
    39,
    198,
    167,
    124,
    67,
    175,
    254,
    1,
    15,
    198,
    62,
    237,
    159,
    31,
    218,
    58,
    244,
    47,
    61,
    67,
    94,
    46,
    78,
    199,
    79,
    89,
    41,
    208,
    135,
    20,
    154,
    35,
    21,
    43,
    230,
    113,
    191,
    147,
    208,
    37,
    28,
    27,
    86,
    203,
    25,
    209,
    148,
    187,
    139,
    111,
    48,
    102,
    76,
    110,
    106,
    130,
    37,
    160,
    34,
    222,
    90,
    165,
    245,
    222,
    102,
    47,
    19,
    130,
    4,
    232,
    137,
    211,
    240,
    11,
    140,
    21
  ],
  B: [
    191,
    154,
    43,
    10,
    207,
    204,
    164,
    231,
    136,
    58,
    239,
    30,
    147,
    230,
    101,
    111,
    150,
    35,
    128,
    57,
    252,
    79,
    173,
    120,
    25,
    126,
    81,
    85,
    8,
    7,
    122,
    237,
    190,
    152,
    246,
    182,
    130,
    219,
    67,
    76,
    167,
    178,
    235,
    250,
    28,
    61,
    186,
    250,
    199,
    67,
    58,
    50,
    86,
    182,
    108,
    77,
    89,
    112,
    59,
    125,
    226,
    50,
    205,
    227,
    125,
    128,
    104,
    27,
    59,
    66,
    53,
    133,
    159,
    203,
    97,
    125,
    139,
    159,
    158,
    7,
    215,
    47,
    140,
    226,
    223,
    231,
    44,
    110,
    184,
    61,
    233,
    47,
    67,
    148,
    22,
    120,
    173,
    156,
    117,
    181,
    94,
    154,
    43,
    10,
    207,
    204,
    164,
    231,
    136,
    58,
    239,
    30,
    147,
    230,
    101,
    111,
    150,
    35,
    128,
    57,
    252,
    79,
    173,
    120,
    25,
    126,
    81,
    85,
    8,
    7,
    122,
    237,
    190,
    152,
    246,
    182,
    130,
    219,
    67,
    76,
    167,
    178,
    235,
    250,
    28,
    61,
    186,
    250,
    199,
    67,
    58,
    50,
    86,
    182,
    108,
    77,
    89,
    112,
    59,
    125,
    226,
    50,
    205,
    227,
    125,
    128,
    104,
    27,
    59,
    66,
    53,
    133,
    159,
    203,
    97,
    125,
    139,
    159,
    158,
    7,
    215,
    47,
    140,
    226,
    223,
    231,
    44,
    110,
    184,
    61,
    233,
    47,
    67,
    148,
    22,
    120,
    173,
    156,
    117,
    181,
    94,
    154,
    43,
    10,
    207,
    204,
    164,
    231,
    136,
    58,
    239,
    30,
    147,
    230,
    101,
    111,
    150,
    35,
    128,
    57,
    252,
    79,
    173,
    120,
    25,
    126,
    81,
    85,
    8,
    7,
    122,
    237,
    190,
    152,
    246,
    182,
    130,
    219,
    67,
    76,
    167,
    178,
    235,
    250,
    28,
    61,
    186,
    250,
    199,
    67,
    58,
    50,
    86,
    182,
    108,
    77
  ],
  A: [
    0,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64
  ],
  I: [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30,
    31,
    32,
    33,
    34,
    35,
    36,
    37,
    38,
    39,
    40,
    41,
    42,
    43,
    44,
    45,
    46,
    47,
    48,
    49,
    50,
    51,
    52,
    53,
    54,
    55,
    56,
    57,
    58,
    59,
    60,
    61,
    62,
    63,
    64,
    65,
    66,
    67,
    68,
    69,
    70,
    71,
    72,
    73,
    74,
    75,
    76,
    77,
    78,
    79,
    80,
    81,
    82,
    83,
    84,
    85,
    86,
    87,
    88,
    89,
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
    99,
    100,
    101,
    102,
    103,
    104,
    105,
    106,
    107,
    108,
    109,
    110,
    111,
    112,
    113,
    114,
    115,
    116,
    117,
    118,
    119,
    120,
    121,
    122,
    123,
    124,
    125,
    126,
    127,
    128,
    129,
    130,
    131,
    132,
    133,
    134,
    135,
    136,
    137,
    138,
    139,
    140,
    141,
    142,
    143,
    144,
    145,
    146,
    147,
    148,
    149,
    150,
    151,
    152,
    153,
    154,
    155,
    156,
    157,
    158,
    159,
    160,
    161,
    162,
    163,
    164,
    165,
    166,
    167,
    168,
    169,
    170,
    171,
    172,
    173,
    174,
    175,
    176,
    177,
    178,
    179,
    180,
    181,
    182,
    183,
    184,
    185,
    186,
    187,
    188,
    189,
    190,
    191,
    192,
    193,
    194,
    195,
    196,
    197,
    198,
    199,
    200,
    201,
    202,
    203,
    204,
    205,
    206,
    207,
    208,
    209,
    210,
    211,
    212,
    213,
    214,
    215,
    216,
    217,
    218,
    219,
    220,
    221,
    222,
    223,
    224,
    225,
    226,
    227,
    228,
    229,
    230,
    231,
    232,
    233,
    234,
    235,
    236,
    237,
    238,
    239,
    240,
    241,
    242,
    243,
    244,
    245,
    246,
    247,
    248,
    249,
    250,
    251,
    252,
    253,
    254,
    255
  ]
};

// src/cmaps/red.json
var red_default = {
  R: [0, 128, 255],
  G: [0, 0, 0],
  B: [0, 0, 0],
  A: [0, 64, 128],
  I: [0, 128, 255]
};

// src/cmaps/redyell.json
var redyell_default = {
  R: [192, 224, 255],
  G: [1, 128, 255],
  B: [0, 0, 0],
  A: [0, 64, 128],
  I: [0, 128, 255]
};

// src/cmaps/rocket.json
var rocket_default = {
  R: [3, 112, 144, 188, 236, 246, 255],
  G: [5, 31, 29, 22, 76, 158, 250],
  B: [26, 87, 91, 86, 62, 117, 235],
  A: [0, 30, 38, 49, 67, 85, 107],
  I: [0, 73, 92, 118, 160, 205, 255]
};

// src/cmaps/surface.json
var surface_default = {
  R: [1, 240, 255],
  G: [1, 128, 255],
  B: [1, 128, 255],
  A: [0, 76, 128],
  I: [0, 153, 255]
};

// src/cmaps/thermal.json
var thermal_default = {
  R: [0, 5, 18, 34, 53, 72, 90, 107, 126, 144, 161, 178, 194, 209, 222, 233, 242, 248, 251, 251, 249, 244, 241, 252],
  G: [0, 4, 10, 11, 9, 11, 17, 23, 30, 36, 43, 50, 59, 70, 83, 98, 117, 136, 157, 177, 199, 220, 240, 254],
  B: [3, 24, 50, 76, 96, 106, 109, 110, 108, 104, 97, 89, 79, 67, 55, 42, 26, 12, 6, 22, 47, 79, 121, 164],
  A: [0, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64],
  I: [0, 11, 22, 33, 44, 55, 66, 77, 88, 99, 110, 121, 133, 144, 155, 166, 177, 188, 199, 210, 221, 232, 243, 255]
};

// src/cmaps/turbo.json
var turbo_default = {
  R: [48, 48, 64, 70, 65, 25, 132, 195, 244, 254, 218, 122],
  G: [18, 18, 64, 107, 150, 226, 255, 241, 199, 158, 57, 4],
  B: [59, 59, 162, 227, 255, 187, 81, 52, 58, 47, 7, 3],
  A: [0, 22, 26, 30, 34, 43, 52, 57, 63, 67, 77, 86],
  I: [0, 1, 16, 32, 49, 83, 118, 140, 164, 181, 219, 255]
};

// src/cmaps/violet.json
var violet_default = {
  R: [0, 128, 255],
  G: [0, 0, 0],
  B: [0, 128, 255],
  A: [0, 64, 128],
  I: [0, 128, 255]
};

// src/cmaps/viridis.json
var viridis_default = {
  R: [68, 49, 53, 253],
  G: [1, 104, 183, 231],
  B: [84, 142, 121, 37],
  A: [0, 56, 80, 88],
  I: [0, 64, 192, 255]
};

// src/cmaps/warm.json
var warm_default = {
  R: [255, 255, 255],
  G: [127, 196, 254],
  B: [0, 0, 0],
  A: [0, 64, 128],
  I: [0, 128, 255]
};

// src/cmaps/winter.json
var winter_default = {
  R: [0, 0, 0],
  G: [0, 128, 255],
  B: [255, 196, 128],
  A: [0, 64, 128],
  I: [0, 128, 255]
};

// src/cmaps/x_rain.json
var x_rain_default = {
  R: [3, 64, 0, 0, 255, 255, 255],
  G: [0, 0, 0, 255, 255, 192, 3],
  B: [0, 32, 48, 56, 64, 96, 128],
  A: [0, 8, 16, 24, 32, 52, 80],
  I: [0, 32, 64, 96, 160, 192, 255]
};

// src/cmaps/afni_blues_inv.json
var afni_blues_inv_default = {
  min: 0,
  max: 0,
  R: [0, 37],
  G: [242, 0],
  B: [255, 255],
  A: [0, 64],
  I: [0, 255]
};

// src/cmaps/afni_reds_inv.json
var afni_reds_inv_default = {
  R: [255, 255],
  G: [255, 11],
  B: [0, 0],
  A: [0, 64],
  I: [0, 255]
};

// src/cmaps/roi_i256.json
var roi_i256_default = {
  min: 0,
  max: 0,
  R: [
    65,
    10,
    223,
    120,
    216,
    207,
    251,
    93,
    252,
    217,
    21,
    253,
    131,
    165,
    173,
    143,
    144,
    217,
    86,
    89,
    63,
    254,
    172,
    1,
    142,
    107,
    42,
    78,
    221,
    8,
    222,
    71,
    127,
    126,
    167,
    33,
    104,
    2,
    208,
    216,
    0,
    251,
    7,
    105,
    2,
    101,
    248,
    190,
    253,
    62,
    255,
    224,
    166,
    37,
    110,
    138,
    45,
    34,
    6,
    37,
    69,
    201,
    43,
    122,
    199,
    37,
    173,
    194,
    103,
    211,
    75,
    159,
    96,
    4,
    239,
    98,
    110,
    193,
    253,
    166,
    40,
    255,
    48,
    130,
    140,
    243,
    101,
    9,
    177,
    220,
    133,
    32,
    4,
    81,
    48,
    48,
    210,
    109,
    60,
    132,
    1,
    119,
    1,
    159,
    247,
    33,
    212,
    187,
    253,
    144,
    196,
    254,
    168,
    79,
    64,
    196,
    39,
    97,
    73,
    173,
    104,
    216,
    217,
    43,
    101,
    119,
    254,
    5,
    237,
    103,
    203,
    122,
    57,
    87,
    251,
    164,
    19,
    75,
    200,
    2,
    252,
    150,
    66,
    0,
    255,
    157,
    23,
    254,
    55,
    16,
    240,
    161,
    69,
    253,
    207,
    195,
    5,
    60,
    255,
    251,
    200,
    217,
    134,
    123,
    253,
    180,
    27,
    246,
    130,
    136,
    250,
    232,
    4,
    125,
    140,
    22,
    253,
    255,
    13,
    180,
    123,
    61,
    254,
    111,
    10,
    185,
    76,
    192,
    255,
    223,
    186,
    61,
    198,
    5,
    172,
    13,
    83,
    172,
    171,
    6,
    23,
    73,
    134,
    133,
    109,
    61,
    213,
    55,
    57,
    132,
    36,
    209,
    2,
    144,
    1,
    253,
    68,
    155,
    3,
    160,
    2,
    77,
    121,
    70,
    67,
    176,
    223,
    131,
    4,
    162,
    232,
    255,
    150,
    94,
    235,
    191,
    207,
    10,
    246,
    0,
    225,
    4,
    209,
    116,
    57,
    112,
    172,
    253,
    1,
    6,
    92,
    227,
    73,
    62,
    135,
    223
  ],
  G: [
    162,
    50,
    112,
    245,
    4,
    124,
    176,
    243,
    56,
    125,
    183,
    139,
    182,
    68,
    189,
    4,
    225,
    10,
    20,
    223,
    7,
    46,
    31,
    193,
    225,
    250,
    219,
    41,
    2,
    100,
    83,
    181,
    34,
    150,
    72,
    223,
    168,
    248,
    80,
    253,
    7,
    117,
    214,
    2,
    248,
    10,
    3,
    59,
    179,
    160,
    90,
    244,
    146,
    4,
    207,
    4,
    125,
    255,
    96,
    100,
    1,
    208,
    130,
    51,
    116,
    181,
    236,
    53,
    244,
    5,
    165,
    28,
    181,
    86,
    96,
    42,
    254,
    1,
    134,
    194,
    214,
    92,
    41,
    204,
    137,
    86,
    207,
    3,
    143,
    3,
    100,
    239,
    164,
    94,
    65,
    251,
    131,
    244,
    173,
    57,
    143,
    107,
    217,
    53,
    210,
    0,
    148,
    250,
    45,
    198,
    81,
    40,
    61,
    218,
    252,
    88,
    171,
    0,
    248,
    24,
    248,
    4,
    41,
    100,
    187,
    46,
    208,
    145,
    43,
    168,
    121,
    46,
    175,
    250,
    125,
    107,
    233,
    112,
    207,
    231,
    174,
    215,
    53,
    9,
    43,
    153,
    52,
    124,
    128,
    65,
    210,
    147,
    255,
    147,
    63,
    200,
    148,
    254,
    0,
    179,
    241,
    42,
    239,
    2,
    230,
    8,
    82,
    135,
    87,
    16,
    3,
    87,
    86,
    151,
    254,
    8,
    255,
    86,
    53,
    19,
    7,
    192,
    171,
    201,
    253,
    247,
    197,
    103,
    251,
    126,
    0,
    149,
    54,
    183,
    61,
    126,
    79,
    113,
    10,
    103,
    184,
    75,
    11,
    195,
    222,
    136,
    149,
    131,
    8,
    99,
    240,
    177,
    252,
    255,
    198,
    16,
    7,
    68,
    178,
    66,
    191,
    150,
    73,
    26,
    211,
    109,
    78,
    209,
    240,
    254,
    1,
    166,
    247,
    131,
    2,
    0,
    167,
    127,
    133,
    10,
    43,
    99,
    235,
    3,
    214,
    142,
    176,
    82,
    132,
    38,
    10,
    249,
    255,
    215,
    44,
    81
  ],
  B: [
    176,
    182,
    248,
    37,
    117,
    35,
    96,
    69,
    32,
    152,
    108,
    20,
    237,
    250,
    2,
    89,
    141,
    216,
    111,
    251,
    211,
    149,
    10,
    44,
    214,
    103,
    31,
    251,
    3,
    32,
    252,
    246,
    97,
    1,
    219,
    167,
    197,
    4,
    36,
    116,
    206,
    118,
    106,
    43,
    205,
    204,
    114,
    69,
    127,
    205,
    87,
    80,
    41,
    251,
    145,
    204,
    253,
    161,
    247,
    1,
    149,
    9,
    43,
    253,
    97,
    72,
    136,
    161,
    171,
    181,
    26,
    255,
    108,
    80,
    218,
    214,
    231,
    255,
    84,
    31,
    109,
    4,
    218,
    3,
    217,
    36,
    68,
    85,
    241,
    39,
    221,
    2,
    240,
    2,
    173,
    42,
    206,
    5,
    110,
    46,
    103,
    27,
    212,
    184,
    2,
    207,
    246,
    45,
    116,
    72,
    110,
    253,
    38,
    105,
    248,
    159,
    243,
    81,
    192,
    93,
    141,
    145,
    24,
    157,
    234,
    131,
    57,
    178,
    62,
    75,
    65,
    176,
    148,
    40,
    253,
    66,
    76,
    240,
    51,
    154,
    17,
    251,
    139,
    253,
    207,
    9,
    114,
    49,
    200,
    254,
    96,
    73,
    138,
    118,
    204,
    102,
    137,
    89,
    145,
    161,
    4,
    112,
    66,
    234,
    147,
    178,
    212,
    205,
    185,
    11,
    203,
    131,
    2,
    250,
    118,
    169,
    1,
    185,
    154,
    53,
    171,
    197,
    61,
    175,
    249,
    96,
    15,
    254,
    95,
    5,
    222,
    75,
    246,
    194,
    2,
    61,
    180,
    25,
    133,
    165,
    15,
    233,
    59,
    35,
    221,
    140,
    109,
    7,
    114,
    255,
    198,
    0,
    115,
    168,
    252,
    23,
    242,
    80,
    75,
    142,
    137,
    255,
    12,
    182,
    68,
    201,
    4,
    111,
    37,
    228,
    83,
    248,
    24,
    192,
    249,
    5,
    54,
    223,
    160,
    122,
    160,
    114,
    145,
    119,
    252,
    31,
    253,
    250,
    10,
    214,
    8,
    47,
    0,
    142,
    222,
    70
  ],
  A: [
    0,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64
  ],
  I: [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
    30,
    31,
    32,
    33,
    34,
    35,
    36,
    37,
    38,
    39,
    40,
    41,
    42,
    43,
    44,
    45,
    46,
    47,
    48,
    49,
    50,
    51,
    52,
    53,
    54,
    55,
    56,
    57,
    58,
    59,
    60,
    61,
    62,
    63,
    64,
    65,
    66,
    67,
    68,
    69,
    70,
    71,
    72,
    73,
    74,
    75,
    76,
    77,
    78,
    79,
    80,
    81,
    82,
    83,
    84,
    85,
    86,
    87,
    88,
    89,
    90,
    91,
    92,
    93,
    94,
    95,
    96,
    97,
    98,
    99,
    100,
    101,
    102,
    103,
    104,
    105,
    106,
    107,
    108,
    109,
    110,
    111,
    112,
    113,
    114,
    115,
    116,
    117,
    118,
    119,
    120,
    121,
    122,
    123,
    124,
    125,
    126,
    127,
    128,
    129,
    130,
    131,
    132,
    133,
    134,
    135,
    136,
    137,
    138,
    139,
    140,
    141,
    142,
    143,
    144,
    145,
    146,
    147,
    148,
    149,
    150,
    151,
    152,
    153,
    154,
    155,
    156,
    157,
    158,
    159,
    160,
    161,
    162,
    163,
    164,
    165,
    166,
    167,
    168,
    169,
    170,
    171,
    172,
    173,
    174,
    175,
    176,
    177,
    178,
    179,
    180,
    181,
    182,
    183,
    184,
    185,
    186,
    187,
    188,
    189,
    190,
    191,
    192,
    193,
    194,
    195,
    196,
    197,
    198,
    199,
    200,
    201,
    202,
    203,
    204,
    205,
    206,
    207,
    208,
    209,
    210,
    211,
    212,
    213,
    214,
    215,
    216,
    217,
    218,
    219,
    220,
    221,
    222,
    223,
    224,
    225,
    226,
    227,
    228,
    229,
    230,
    231,
    232,
    233,
    234,
    235,
    236,
    237,
    238,
    239,
    240,
    241,
    242,
    243,
    244,
    245,
    246,
    247,
    248,
    249,
    250,
    251,
    252,
    253,
    254,
    255
  ]
};

// src/colortables.ts
var ColorTables = class {
  /**
   * Sets cluts to alphabetically sorted cmaps
   */
  constructor() {
    __publicField(this, "gamma", 1);
    __publicField(this, "version", 0.1);
    __publicField(this, "cluts", {});
    const cmapKeys = Object.keys(cmaps_exports);
    const cmapsSorted = cmapKeys.filter((k) => !k.startsWith("$")).sort(new Intl.Collator("en").compare);
    for (const key of cmapsSorted) {
      this.cluts[key] = cmaps_exports[key];
    }
  }
  addColormap(key, cmap) {
    this.cluts[key] = cmap;
  }
  colormaps() {
    return Object.keys(this.cluts);
  }
  // for backward compatibility: prior to v0.34 "colormaps" used to be "colorMaps"
  colorMaps() {
    return this.colormaps();
  }
  // returns key name if it exists, otherwise returns default "gray"
  colormapFromKey(name) {
    let cmap = this.cluts[name];
    if (cmap !== void 0) {
      return cmap;
    }
    cmap = this.cluts[name.toLowerCase()];
    if (cmap !== void 0) {
      return cmap;
    }
    if (name.length > 0) {
      log.warn("No color map named " + name);
    }
    return {
      min: 0,
      max: 0,
      R: [0, 255],
      G: [0, 255],
      B: [0, 255],
      A: [0, 255],
      I: [0, 255]
    };
  }
  // not included in public docs
  colormap(key = "", isInvert = false) {
    const cmap = this.colormapFromKey(key);
    return this.makeLut(cmap.R, cmap.G, cmap.B, cmap.A, cmap.I, isInvert);
  }
  makeLabelLut(cm, alphaFill = 255, maxIdx = Infinity) {
    if (cm.R === void 0 || cm.G === void 0 || cm.B === void 0) {
      throw new Error(`Invalid colormap table: ${cm}`);
    }
    const nLabels = cm.R.length;
    const idxs = cm.I ?? [...Array(nLabels).keys()];
    let hasInvalid = false;
    for (let i = 0; i < idxs.length; i++) {
      if (idxs[i] > maxIdx) {
        hasInvalid = true;
        idxs[i] = maxIdx;
      }
    }
    if (hasInvalid) {
      log.warn(`Some colormap indices clamped to match label range.`);
    }
    if (nLabels !== cm.G.length || nLabels !== cm.B.length || nLabels !== idxs.length) {
      throw new Error(`colormap does not make sense: ${cm} Rs ${cm.R.length} Gs ${cm.G.length} Bs ${cm.B.length} Is ${idxs.length}`);
    }
    let As = new Uint8ClampedArray(nLabels).fill(alphaFill);
    const zeroPos = idxs.indexOf(0);
    if (zeroPos >= 0) {
      As[zeroPos] = 0;
    }
    if (cm.A !== void 0) {
      As = Uint8ClampedArray.from(cm.A);
    }
    const mnIdx = Math.min(...idxs);
    const mxIdx = Math.max(...idxs);
    const nLabelsDense = mxIdx - mnIdx + 1;
    const lut = new Uint8ClampedArray(nLabelsDense * 4).fill(0);
    for (let i = 0; i < nLabels; i++) {
      let k = (idxs[i] - mnIdx) * 4;
      lut[k++] = cm.R[i];
      lut[k++] = cm.G[i];
      lut[k++] = cm.B[i];
      lut[k++] = As[i];
    }
    const cmap = {
      lut,
      min: mnIdx,
      max: mxIdx
    };
    if (cm.labels) {
      const nL = cm.labels.length;
      if (nL === nLabelsDense) {
        cmap.labels = cm.labels;
      } else if (nL === nLabels) {
        cmap.labels = Array(nLabelsDense).fill("?");
        for (let i = 0; i < nLabels; i++) {
          cmap.labels[idxs[i] - mnIdx] = cm.labels[i];
        }
      }
    }
    return cmap;
  }
  async makeLabelLutFromUrl(name, alphaFill = 255, maxIdx = Infinity) {
    const response = await fetch(name);
    const cm = await response.json();
    return this.makeLabelLut(cm, alphaFill, maxIdx);
  }
  // not included in public docs
  // The drawing colormap is a variant of the label colormap with precisely 256 colors
  makeDrawLut(name) {
    let cmap = typeof name === "object" ? name : cmaps_exports[name];
    if (cmap === void 0) {
      log.warn("colormap undefined ", name);
      cmap = this.colormapFromKey("");
    }
    const cm = this.makeLabelLut(cmap, 255);
    if (cm.labels === void 0) {
      cm.labels = [];
    }
    if (cm.labels.length < 256) {
      const j = cm.labels.length;
      for (let i = j; i < 256; i++) {
        cm.labels.push(i.toString());
      }
    }
    const lut = new Uint8ClampedArray(256 * 4);
    let k = 0;
    for (let i = 0; i < 256; i++) {
      lut[k++] = 255;
      lut[k++] = 0;
      lut[k++] = 0;
      lut[k++] = 255;
    }
    lut[3] = 0;
    const explicitLUTbytes = Math.min(cm.lut.length, 256 * 4);
    if (explicitLUTbytes > 0) {
      for (let i = 0; i < explicitLUTbytes; i++) {
        lut[i] = cm.lut[i];
      }
    }
    return {
      lut,
      labels: cm.labels
    };
  }
  // not included in public docs
  makeLut(Rsi, Gsi, Bsi, Asi, Isi, isInvert) {
    const nIdx = Rsi.length;
    const Rs = [...Rsi];
    const Gs = [...Gsi];
    const Bs = [...Bsi];
    if (!Isi) {
      Isi = new Array(nIdx);
      for (let i = 0; i < nIdx; i++) {
        Isi[i] = i / (nIdx - 1) * 255;
      }
    }
    if (!Asi) {
      Asi = new Array(nIdx).fill(64);
      Asi[0] = 0;
    }
    let As = Uint8ClampedArray.from(Asi);
    let Is = Uint8ClampedArray.from(Isi);
    if (isInvert) {
      for (let i = 0; i < nIdx; i++) {
        Rs[i] = Rsi[nIdx - 1 - i];
        Gs[i] = Gsi[nIdx - 1 - i];
        Bs[i] = Bsi[nIdx - 1 - i];
        As[i] = 255 - Asi[nIdx - 1 - i];
        Is[i] = 255 - Isi[nIdx - 1 - i];
      }
    }
    const lut = new Uint8ClampedArray(256 * 4);
    if (typeof Is === "undefined") {
      Is = new Uint8ClampedArray(nIdx).fill(0);
      for (let i = 0; i < nIdx; i++) {
        Is[i] = Math.round(i * 255 / (nIdx - 1));
      }
    }
    if (typeof As === "undefined") {
      As = new Uint8ClampedArray(nIdx).fill(64);
      As[0] = 0;
    }
    for (let i = 0; i < nIdx - 1; i++) {
      const idxLo = Is[i];
      let idxHi = Is[i + 1];
      if (i === 0 && idxLo !== 0) {
        log.warn("colormap issue: indices expected to start with 0 not ", idxLo);
      }
      if (i === Is.length - 2 && idxHi !== 255) {
        log.warn("padding colormap: indices expected end with 255 not ", idxHi);
        idxHi = 255;
      }
      const idxRng = idxHi - idxLo;
      let k = idxLo * 4;
      for (let j = idxLo; j <= idxHi; j++) {
        const f = (j - idxLo) / idxRng;
        lut[k++] = Rs[i] + f * (Rs[i + 1] - Rs[i]);
        lut[k++] = Gs[i] + f * (Gs[i + 1] - Gs[i]);
        lut[k++] = Bs[i] + f * (Bs[i + 1] - Bs[i]);
        lut[k++] = As[i] + f * (As[i + 1] - As[i]);
      }
    }
    if (this.gamma === 1) {
      return lut;
    }
    for (let i = 0; i < 255 * 4; i++) {
      if (i % 4 === 3) {
        continue;
      }
      lut[i] = Math.pow(lut[i] / 255, 1 / this.gamma) * 255;
    }
    return lut;
  }
};
var cmapper = new ColorTables();

// src/nvimage/ImageFactory.ts
import { Gunzip } from "fflate";
import { readHeaderAsync as readHeaderAsync3 } from "nifti-reader-js";

// src/nvimage/ZarrProcessor.ts
import * as zarr from "zarrita";

// src/nvimage/zarr/NVZarrHelper.ts
import { NIFTI1 as NIFTI114 } from "nifti-reader-js";
import { v4 as uuidv4 } from "@lukeed/uuid";

// src/nvimage/affineUtils.ts
import { mat4 as mat47 } from "gl-matrix";

// src/nvimage/zarr/ZarrChunkClient.ts
import * as zarr2 from "zarrita";

// src/nvmesh.ts
import { vec3 as vec311, vec4 as vec48 } from "gl-matrix";
import { v4 as uuidv43 } from "@lukeed/uuid";

// src/niivue-object3D.ts
import { mat4 as mat48, vec3 as vec38 } from "gl-matrix";

// src/nvmesh-utilities.ts
import { vec3 as vec39 } from "gl-matrix";

// src/nvmesh-loaders.ts
import { mat4 as mat49, vec4 as vec47, vec3 as vec310 } from "gl-matrix";

// src/nvmesh.ts
var NVMeshLayerDefaults = {
  colormap: "gray",
  opacity: 0,
  nFrame4D: 0,
  frame4D: 0,
  outlineBorder: 0,
  cal_min: 0,
  cal_max: 0,
  cal_minNeg: 0,
  cal_maxNeg: 0,
  colormapType: 0 /* MIN_TO_MAX */,
  values: new Array(),
  useNegativeCmap: false,
  showLegend: true
};

// src/utils/legacy-migrate.ts
import { deserialize } from "@ungap/structured-clone";

// src/nvdocument.ts
var DEFAULT_OPTIONS = {
  // ... (same defaults as your original file)
  textHeight: -1,
  fontSizeScaling: 0.4,
  fontMinPx: 13,
  colorbarHeight: 0.05,
  colorbarWidth: -1,
  showColorbarBorder: true,
  crosshairWidth: 1,
  crosshairWidthUnit: "voxels",
  crosshairGap: 0,
  rulerWidth: 4,
  show3Dcrosshair: false,
  backColor: [0, 0, 0, 1],
  crosshairColor: [1, 0, 0, 1],
  fontColor: [0.5, 0.5, 0.5, 1],
  selectionBoxColor: [1, 1, 1, 0.5],
  clipPlaneColor: [0.7, 0, 0.7, 0.5],
  isClipPlanesCutaway: false,
  isClipAllVolumes: false,
  paqdUniforms: [0.3, 0.5, 0.5, 1],
  rulerColor: [1, 0, 0, 0.8],
  colorbarMargin: 0.05,
  trustCalMinMax: true,
  clipPlaneHotKey: "KeyC",
  cycleClipPlaneHotKey: "KeyP",
  viewModeHotKey: "KeyV",
  doubleTouchTimeout: 500,
  longTouchTimeout: 1e3,
  keyDebounceTime: 50,
  isNearestInterpolation: false,
  isResizeCanvas: true,
  atlasOutline: 0,
  atlasActiveIndex: 0,
  isRuler: false,
  isColorbar: false,
  isOrientCube: false,
  tileMargin: 0,
  multiplanarPadPixels: 0,
  multiplanarForceRender: false,
  multiplanarEqualSize: false,
  multiplanarShowRender: 2 /* AUTO */,
  isRadiologicalConvention: false,
  meshThicknessOn2D: Infinity,
  dragMode: 1 /* contrast */,
  dragModePrimary: 8 /* crosshair */,
  mouseEventConfig: void 0,
  touchEventConfig: void 0,
  yoke3Dto2DZoom: false,
  isDepthPickMesh: false,
  isCornerOrientationText: false,
  isOrientationTextVisible: true,
  showAllOrientationMarkers: false,
  heroImageFraction: 0,
  heroSliceType: 4 /* RENDER */,
  sagittalNoseLeft: false,
  isSliceMM: false,
  isV1SliceShader: false,
  forceDevicePixelRatio: 0,
  logLevel: "info",
  loadingText: "loading ...",
  isForceMouseClickToVoxelCenters: false,
  dragAndDropEnabled: true,
  drawingEnabled: false,
  penValue: 1,
  penType: 0 /* PEN */,
  floodFillNeighbors: 6,
  isFilledPen: false,
  thumbnail: "",
  maxDrawUndoBitmaps: 8,
  sliceType: 3 /* MULTIPLANAR */,
  meshXRay: 0,
  isAntiAlias: null,
  limitFrames4D: NaN,
  isAdditiveBlend: false,
  showLegend: true,
  legendBackgroundColor: [0.3, 0.3, 0.3, 0.5],
  legendTextColor: [1, 1, 1, 1],
  multiplanarLayout: 0 /* AUTO */,
  renderOverlayBlend: 1,
  sliceMosaicString: "",
  centerMosaic: false,
  penSize: 1,
  interactive: true,
  clickToSegment: false,
  clickToSegmentRadius: 3,
  clickToSegmentBright: true,
  clickToSegmentAutoIntensity: false,
  clickToSegmentIntensityMax: NaN,
  clickToSegmentIntensityMin: NaN,
  clickToSegmentPercent: 0,
  clickToSegmentMaxDistanceMM: Number.POSITIVE_INFINITY,
  clickToSegmentIs2D: false,
  selectionBoxLineThickness: 4,
  selectionBoxIsOutline: false,
  scrollRequiresFocus: false,
  showMeasureUnits: true,
  measureTextJustify: "center",
  measureTextColor: [1, 0, 0, 1],
  measureLineColor: [1, 0, 0, 1],
  measureTextHeight: 0.06,
  isAlphaClipDark: false,
  gradientOrder: 1,
  gradientOpacity: 0,
  renderSilhouette: 0,
  gradientAmount: 0,
  invertScrollDirection: false,
  is2DSliceShader: false,
  bounds: null,
  showBoundsBorder: false,
  boundsBorderColor: [1, 1, 1, 1],
  // white border by default
  windowingGainFactor: 2,
  // Zarr options
  zarrCacheSize: 1e3,
  zarrPrefetchRings: 10,
  smoothDrawing: 0
};
var INITIAL_SCENE_DATA = {
  gamma: 1,
  azimuth: 110,
  elevation: 10,
  crosshairPos: vec313.fromValues(0.5, 0.5, 0.5),
  clipPlanes: [[0, 0, 0, 0]],
  clipPlaneDepthAziElevs: [[2, 0, 0]],
  volScaleMultiplier: 1,
  pan2Dxyzmm: vec49.fromValues(0, 0, 0, 1)
};

// src/drawing/masks.ts
function findBoundarySlices(sliceType, drawBitmap, dims) {
  const { dimX, dimY, dimZ } = dims;
  let axisSize;
  if (sliceType === 0 /* AXIAL */) {
    axisSize = dimZ;
  } else if (sliceType === 1 /* CORONAL */) {
    axisSize = dimY;
  } else if (sliceType === 2 /* SAGITTAL */) {
    axisSize = dimX;
  } else {
    return null;
  }
  let firstSliceWithData = -1;
  let lastSliceWithData = -1;
  for (let slice3 = 0; slice3 < axisSize; slice3++) {
    let hasData = false;
    if (sliceType === 0 /* AXIAL */) {
      const offset = slice3 * dimX * dimY;
      for (let i = 0; i < dimX * dimY; i++) {
        if (drawBitmap[offset + i] > 0) {
          hasData = true;
          break;
        }
      }
    } else if (sliceType === 1 /* CORONAL */) {
      for (let z = 0; z < dimZ; z++) {
        for (let x = 0; x < dimX; x++) {
          const idx = x + slice3 * dimX + z * dimX * dimY;
          if (drawBitmap[idx] > 0) {
            hasData = true;
            break;
          }
        }
        if (hasData) {
          break;
        }
      }
    } else if (sliceType === 2 /* SAGITTAL */) {
      for (let z = 0; z < dimZ; z++) {
        for (let y = 0; y < dimY; y++) {
          const idx = slice3 + y * dimX + z * dimX * dimY;
          if (drawBitmap[idx] > 0) {
            hasData = true;
            break;
          }
        }
        if (hasData) {
          break;
        }
      }
    }
    if (hasData) {
      if (firstSliceWithData === -1) {
        firstSliceWithData = slice3;
      }
      lastSliceWithData = slice3;
    }
  }
  if (firstSliceWithData === -1 || lastSliceWithData === -1) {
    return null;
  }
  return { first: firstSliceWithData, last: lastSliceWithData };
}
function extractSlice(sliceIndex, sliceType, drawBitmap, dims) {
  const { dimX, dimY, dimZ } = dims;
  let sliceData;
  if (sliceType === 0 /* AXIAL */) {
    sliceData = new Float32Array(dimX * dimY);
    const offset = sliceIndex * dimX * dimY;
    for (let i = 0; i < dimX * dimY; i++) {
      sliceData[i] = drawBitmap[offset + i];
    }
  } else if (sliceType === 1 /* CORONAL */) {
    sliceData = new Float32Array(dimX * dimZ);
    for (let z = 0; z < dimZ; z++) {
      for (let x = 0; x < dimX; x++) {
        const srcIdx = x + sliceIndex * dimX + z * dimX * dimY;
        const dstIdx = x + z * dimX;
        sliceData[dstIdx] = drawBitmap[srcIdx];
      }
    }
  } else if (sliceType === 2 /* SAGITTAL */) {
    sliceData = new Float32Array(dimY * dimZ);
    for (let z = 0; z < dimZ; z++) {
      for (let y = 0; y < dimY; y++) {
        const srcIdx = sliceIndex + y * dimX + z * dimX * dimY;
        const dstIdx = y + z * dimY;
        sliceData[dstIdx] = drawBitmap[srcIdx];
      }
    }
  } else {
    throw new Error("Invalid slice type");
  }
  return sliceData;
}
function extractIntensitySlice(sliceIndex, sliceType, imageData, dims, maxVal) {
  const { dimX, dimY, dimZ } = dims;
  let sliceData;
  if (sliceType === 0 /* AXIAL */) {
    sliceData = new Float32Array(dimX * dimY);
    const offset = sliceIndex * dimX * dimY;
    for (let i = 0; i < dimX * dimY; i++) {
      sliceData[i] = imageData[offset + i] / maxVal;
    }
  } else if (sliceType === 1 /* CORONAL */) {
    sliceData = new Float32Array(dimX * dimZ);
    for (let z = 0; z < dimZ; z++) {
      for (let x = 0; x < dimX; x++) {
        const srcIdx = x + sliceIndex * dimX + z * dimX * dimY;
        const dstIdx = x + z * dimX;
        sliceData[dstIdx] = imageData[srcIdx] / maxVal;
      }
    }
  } else if (sliceType === 2 /* SAGITTAL */) {
    sliceData = new Float32Array(dimY * dimZ);
    for (let z = 0; z < dimZ; z++) {
      for (let y = 0; y < dimY; y++) {
        const srcIdx = sliceIndex + y * dimX + z * dimX * dimY;
        const dstIdx = y + z * dimY;
        sliceData[dstIdx] = imageData[srcIdx] / maxVal;
      }
    }
  } else {
    throw new Error("Invalid slice type");
  }
  return sliceData;
}
function insertColorMask(mask, sliceIndex, sliceType, drawBitmap, dims, binaryThreshold, color) {
  const { dimX, dimY, dimZ } = dims;
  if (sliceType === 0 /* AXIAL */) {
    const offset = sliceIndex * dimX * dimY;
    for (let i = 0; i < mask.length; i++) {
      if (mask[i] >= binaryThreshold) {
        drawBitmap[offset + i] = color;
      }
    }
  } else if (sliceType === 1 /* CORONAL */) {
    for (let z = 0; z < dimZ; z++) {
      for (let x = 0; x < dimX; x++) {
        const srcIdx = x + z * dimX;
        const dstIdx = x + sliceIndex * dimX + z * dimX * dimY;
        if (mask[srcIdx] >= binaryThreshold) {
          drawBitmap[dstIdx] = color;
        }
      }
    }
  } else if (sliceType === 2 /* SAGITTAL */) {
    for (let z = 0; z < dimZ; z++) {
      for (let y = 0; y < dimY; y++) {
        const srcIdx = y + z * dimY;
        const dstIdx = sliceIndex + y * dimX + z * dimX * dimY;
        if (mask[srcIdx] >= binaryThreshold) {
          drawBitmap[dstIdx] = color;
        }
      }
    }
  } else {
    throw new Error("Invalid slice type");
  }
}
function smoothSlice(slice3, width, height) {
  if (width < 3 || height < 3) {
    return;
  }
  const temp = new Float32Array(slice3.length);
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const idx = x + y * width;
      if (x === 0 || x === width - 1) {
        temp[idx] = slice3[idx];
      } else {
        temp[idx] = (slice3[idx - 1] + 2 * slice3[idx] + slice3[idx + 1]) * 0.25;
      }
    }
  }
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const idx = x + y * width;
      if (y === 0 || y === height - 1) {
        slice3[idx] = temp[idx];
      } else {
        slice3[idx] = (temp[idx - width] + 2 * temp[idx] + temp[idx + width]) * 0.25;
      }
    }
  }
}
function calculateIntensityWeight(intensity1, intensity2, targetIntensity, intensitySigma) {
  const diff1 = Math.abs(targetIntensity - intensity1);
  const diff2 = Math.abs(targetIntensity - intensity2);
  const weight1 = Math.exp(-diff1 * diff1 / (2 * intensitySigma * intensitySigma));
  const weight2 = Math.exp(-diff2 * diff2 / (2 * intensitySigma * intensitySigma));
  const totalWeight = weight1 + weight2;
  if (totalWeight < 1e-6) {
    return 0.5;
  }
  return weight1 / totalWeight;
}
function doGeometricInterpolation(sliceLow, sliceHigh, z, sliceIndexLow, sliceIndexHigh, interpolatedSlice) {
  const fracHigh = (z - sliceIndexLow) / (sliceIndexHigh - sliceIndexLow);
  const fracLow = 1 - fracHigh;
  for (let i = 0; i < sliceLow.length; i++) {
    interpolatedSlice[i] = sliceLow[i] * fracLow + sliceHigh[i] * fracHigh;
  }
}
function doIntensityGuidedInterpolation(sliceLow, sliceHigh, z, sliceIndexLow, sliceIndexHigh, interpolatedSlice, opts, intensityLow, intensityHigh, targetIntensity) {
  const baseFracHigh = (z - sliceIndexLow) / (sliceIndexHigh - sliceIndexLow);
  const baseFracLow = 1 - baseFracHigh;
  for (let i = 0; i < sliceLow.length; i++) {
    if (sliceLow[i] > 0 || sliceHigh[i] > 0) {
      const intensityWeight = calculateIntensityWeight(intensityLow[i], intensityHigh[i], targetIntensity[i], opts.intensitySigma);
      const alpha = opts.intensityWeight;
      const combinedWeightLow = alpha * intensityWeight + (1 - alpha) * baseFracLow;
      const combinedWeightHigh = 1 - combinedWeightLow;
      interpolatedSlice[i] = sliceLow[i] * combinedWeightLow + sliceHigh[i] * combinedWeightHigh;
    } else {
      interpolatedSlice[i] = sliceLow[i] * baseFracLow + sliceHigh[i] * baseFracHigh;
    }
  }
}
function interpolateMaskSlices(drawBitmap, dims, imageData, maxVal, sliceIndexLow, sliceIndexHigh, options, refreshDrawingCallback) {
  const { dimX, dimY, dimZ } = dims;
  const sliceType = options.sliceType ?? 0 /* AXIAL */;
  let sliceWidth, sliceHeight, maxSliceIndex;
  if (sliceType === 0 /* AXIAL */) {
    sliceWidth = dimX;
    sliceHeight = dimY;
    maxSliceIndex = dimZ - 1;
  } else if (sliceType === 1 /* CORONAL */) {
    sliceWidth = dimX;
    sliceHeight = dimZ;
    maxSliceIndex = dimY - 1;
  } else if (sliceType === 2 /* SAGITTAL */) {
    sliceWidth = dimY;
    sliceHeight = dimZ;
    maxSliceIndex = dimX - 1;
  } else {
    throw new Error("Invalid slice type. Must be AXIAL, CORONAL, or SAGITTAL");
  }
  const opts = {
    intensityWeight: options.intensityWeight ?? 0.7,
    binaryThreshold: options.binaryThreshold ?? 0.375,
    intensitySigma: options.intensitySigma ?? 0.1,
    applySmoothingToSlices: options.applySmoothingToSlices ?? true,
    useIntensityGuided: options.useIntensityGuided ?? true
  };
  if (sliceIndexLow !== void 0 && sliceIndexHigh !== void 0) {
    if (sliceIndexLow >= sliceIndexHigh) {
      throw new Error("Low slice index must be less than high slice index");
    }
    if (sliceIndexLow < 0 || sliceIndexHigh > maxSliceIndex) {
      throw new Error(`Slice indices out of bounds [0, ${maxSliceIndex}]`);
    }
  }
  const colorRanges = /* @__PURE__ */ new Map();
  for (let sliceIdx = 0; sliceIdx <= maxSliceIndex; sliceIdx++) {
    const slice3 = extractSlice(sliceIdx, sliceType, drawBitmap, dims);
    for (let i = 0; i < slice3.length; i++) {
      const color = slice3[i];
      if (color > 0) {
        if (!colorRanges.has(color)) {
          colorRanges.set(color, { min: sliceIdx, max: sliceIdx });
        } else {
          const range = colorRanges.get(color);
          range.min = Math.min(range.min, sliceIdx);
          range.max = Math.max(range.max, sliceIdx);
        }
      }
    }
  }
  for (const [color, range] of colorRanges) {
    const colorSliceLow = sliceIndexLow !== void 0 ? Math.max(sliceIndexLow, range.min) : range.min;
    const colorSliceHigh = sliceIndexHigh !== void 0 ? Math.min(sliceIndexHigh, range.max) : range.max;
    if (colorSliceLow >= colorSliceHigh || colorSliceHigh - colorSliceLow < 2) {
      continue;
    }
    const sliceLow = extractSlice(colorSliceLow, sliceType, drawBitmap, dims);
    const sliceHigh = extractSlice(colorSliceHigh, sliceType, drawBitmap, dims);
    const colorMaskLow = new Float32Array(sliceLow.length);
    const colorMaskHigh = new Float32Array(sliceHigh.length);
    for (let i = 0; i < sliceLow.length; i++) {
      colorMaskLow[i] = sliceLow[i] === color ? 1 : 0;
      colorMaskHigh[i] = sliceHigh[i] === color ? 1 : 0;
    }
    if (opts.applySmoothingToSlices) {
      smoothSlice(colorMaskLow, sliceWidth, sliceHeight);
      smoothSlice(colorMaskHigh, sliceWidth, sliceHeight);
    }
    for (let z = colorSliceLow + 1; z < colorSliceHigh; z++) {
      const colorInterpolated = new Float32Array(sliceWidth * sliceHeight);
      if (opts.useIntensityGuided && imageData) {
        const intensityLow = extractIntensitySlice(colorSliceLow, sliceType, imageData, dims, maxVal);
        const intensityHigh = extractIntensitySlice(colorSliceHigh, sliceType, imageData, dims, maxVal);
        const targetIntensity = extractIntensitySlice(z, sliceType, imageData, dims, maxVal);
        doIntensityGuidedInterpolation(colorMaskLow, colorMaskHigh, z, colorSliceLow, colorSliceHigh, colorInterpolated, opts, intensityLow, intensityHigh, targetIntensity);
      } else {
        doGeometricInterpolation(colorMaskLow, colorMaskHigh, z, colorSliceLow, colorSliceHigh, colorInterpolated);
      }
      insertColorMask(colorInterpolated, z, sliceType, drawBitmap, dims, opts.binaryThreshold, color);
    }
  }
  refreshDrawingCallback();
}
export {
  calculateIntensityWeight,
  decodeRLE,
  doGeometricInterpolation,
  doIntensityGuidedInterpolation,
  drawUndo,
  encodeRLE,
  extractIntensitySlice,
  extractSlice,
  findBoundarySlices,
  insertColorMask,
  interpolateMaskSlices,
  smoothSlice
};
